<!DOCTYPE html><html lang="zh-CN" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no"><title>RNA-Seq | 海底捞里没有鱼</title><meta name="keywords" content="转录组"><meta name="author" content="泡泡"><meta name="copyright" content="泡泡"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#f7f9fe"><meta name="mobile-web-app-capable" content="yes"><meta name="apple-touch-fullscreen" content="yes"><meta name="apple-mobile-web-app-title" content="RNA-Seq"><meta name="application-name" content="RNA-Seq"><meta name="apple-mobile-web-app-capable" content="yes"><meta name="apple-mobile-web-app-status-bar-style" content="#f7f9fe"><meta property="og:type" content="article"><meta property="og:title" content="RNA-Seq"><meta property="og:url" content="https://smileszh.github.io/posts/7acd0447.html"><meta property="og:site_name" content="海底捞里没有鱼"><meta property="og:description" content="RNA-Seq指南RNA-Seq参考根据基因组进行量化RNA-Seq根据基因组进行量化时，参考数据是基因组序列，基因组注释文件是基因组的注释信息，包括基因的位置、外显子的位置、基因的功能等信息。该方法将对齐文件与注释相交叉，以产生“计数”，然后过滤这些计数以保留统计上显著的结果。定量矩阵代表对齐与标"><meta property="og:locale" content="zh-CN"><meta property="og:image" content="https://bu.dusays.com/2023/07/03/64a283d519f00.png"><meta property="article:author" content="泡泡"><meta property="article:tag" content="Linux, R, 生物信息学, 笔记"><meta name="twitter:card" content="summary"><meta name="twitter:image" content="https://bu.dusays.com/2023/07/03/64a283d519f00.png"><meta name="description" content="RNA-Seq指南RNA-Seq参考根据基因组进行量化RNA-Seq根据基因组进行量化时，参考数据是基因组序列，基因组注释文件是基因组的注释信息，包括基因的位置、外显子的位置、基因的功能等信息。该方法将对齐文件与注释相交叉，以产生“计数”，然后过滤这些计数以保留统计上显著的结果。定量矩阵代表对齐与标"><link rel="shortcut icon" href="/img/favicon.ico"><link rel="canonical" href="https://smileszh.github.io/posts/7acd0447"><link rel="preconnect" href="//npm.elemecdn.com"/><link rel="preconnect" href="//npm.onmicrosoft.cn"/><link rel="preconnect" href="//hm.baidu.com"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><meta name="google-site-verification" content="C8GeV4w16I6nsJ_CstXPp4zcoh5s3ok1bV93asrNPtg"/><meta name="baidu-site-verification" content="codeva-P3G6GZaKch"/><meta name="msvalidate.01" content="xxx"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.cbd.int/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.cbd.int/@fancyapps/ui@5.0.20/dist/fancybox/fancybox.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://npm.elemecdn.com/anzhiyu-theme-static@1.0.0/swiper/swiper.min.css" media="print" onload="this.media='all'"><script>var _hmt = _hmt || [];
(function() {
  var hm = document.createElement("script");
  hm.src = "https://hm.baidu.com/hm.js?42279978f5aed61b92bf91dc1b8327df";
  var s = document.getElementsByTagName("script")[0]; 
  s.parentNode.insertBefore(hm, s);
})();
</script><script>const GLOBAL_CONFIG = {
  linkPageTop: {"enable":true,"title":"与数百名博主无限进步","addFriendPlaceholder":"昵称（请勿包含博客等字样）：\n网站地址（要求博客地址，请勿提交个人主页）：\n头像图片url（请提供尽可能清晰的图片，我会上传到我自己的图床）：\n描述：\n站点截图（可选）：\n"},
  postHeadAiDescription: {"enable":true,"gptName":"海底捞里没有鱼","mode":"local","switchBtn":false,"btnLink":"https://afdian.net/item/886a79d4db6711eda42a52540025c377","randomNum":3,"basicWordCount":1000,"key":"7MA3Up7FfWRm8MBtP7A1","Referer":"https://www.smileszh.cn/"},
  diytitle: {"enable":true,"leaveTitle":"w(ﾟДﾟ)w 不要走！再看看嘛！","backTitle":"♪(^∇^*)欢迎肥来！"},
  LA51: undefined,
  greetingBox: {"enable":true,"default":"晚上好👋","list":[{"greeting":"晚安😴","startTime":0,"endTime":5},{"greeting":"早上好鸭👋, 祝你一天好心情！","startTime":6,"endTime":9},{"greeting":"上午好👋, 状态很好，鼓励一下～","startTime":10,"endTime":10},{"greeting":"11点多啦, 在坚持一下就吃饭啦～","startTime":11,"endTime":11},{"greeting":"午安👋, 宝贝","startTime":12,"endTime":14},{"greeting":"🌈充实的一天辛苦啦！","startTime":14,"endTime":18},{"greeting":"19点喽, 奖励一顿丰盛的大餐吧🍔。","startTime":19,"endTime":19},{"greeting":"晚上好👋, 在属于自己的时间好好放松😌~","startTime":20,"endTime":24}]},
  twikooEnvId: 'https://twikoo.smileszh.cn/',
  commentBarrageConfig:{"enable":true,"maxBarrage":1,"barrageTime":4000,"accessToken":"209b577e408ea76149a174f2fdccf135","mailMd5":"6b964b5aa12ded0192d89a54488aa4c3"},
  root: '/',
  preloader: {"source":3},
  friends_vue_info: {"apiurl":"https://link.smileszh.cn/"},
  navMusic: true,
  mainTone: undefined,
  authorStatus: {"skills":["🤖️ 数码科技爱好者","🔍 分享与热心帮助","💢 壮汉人狠话不多"]},
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":true,"languages":{"hits_empty":"找不到您查询的内容：${query}"}},
  translate: {"defaultEncoding":2,"translateDelay":0,"msgToTraditionalChinese":"繁","msgToSimplifiedChinese":"简","rightMenuMsgToTraditionalChinese":"转为繁体","rightMenuMsgToSimplifiedChinese":"转为简体"},
  noticeOutdate: {"limitDay":365,"position":"top","messagePrev":"It has been","messageNext":"days since the last update, the content of the article may be outdated."},
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":330},
  copy: {
    success: '复制成功',
    error: '复制错误',
    noSupport: '浏览器不支持'
  },
  relativeDate: {
    homepage: false,
    simplehomepage: true,
    post: false
  },
  runtime: '天',
  date_suffix: {
    just: '刚刚',
    min: '分钟前',
    hour: '小时前',
    day: '天前',
    month: '个月前'
  },
  copyright: {"copy":true,"copyrightEbable":false,"limitCount":50,"languages":{"author":"作者: 泡泡","link":"链接: ","source":"来源: 海底捞里没有鱼","info":"著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。","copySuccess":"复制成功，复制和转载请标注本文地址"}},
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"你已切换为繁体","cht_to_chs":"你已切换为简体","day_to_night":"你已切换为深色模式","night_to_day":"你已切换为浅色模式","bgLight":"#3b70fc","bgDark":"#1f1f1f","position":"top-center"},
  source: {
    justifiedGallery: {
      js: 'https://cdn.cbd.int/flickr-justified-gallery@2.1.2/dist/fjGallery.min.js',
      css: 'https://cdn.cbd.int/flickr-justified-gallery@2.1.2/dist/fjGallery.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: true,
  isAnchor: false,
  shortcutKey: undefined,
  autoDarkmode: true
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  configTitle: '海底捞里没有鱼',
  title: 'RNA-Seq',
  postAI: '',
  pageFillDescription: 'RNA-Seq指南, RNA-Seq参考, 根据基因组进行量化, 根据转录组进行分类, 参考基因组, 参考基因组大小, 有多少个染色体, 每个染色体的长度, 参考基因组注释文件, 了解数据, 比对, 构建索引, 比对, bam文件转换, 计数, 差异表达分析, RNA-Seq分析所遇问题, 比对问题, 计数问题, 链特异性问题, 转录完整性, RNA-Seq分析流程, 基因表达量数据类型, 基因表达量数据转换指南参考根据基因组进行量化根据基因组进行量化时参考数据是基因组序列基因组注释文件是基因组的注释信息包括基因的位置外显子的位置基因的功能等信息该方法将对齐文件与注释相交叉以产生计数然后过滤这些计数以保留统计上显著的结果定量矩阵代表对齐与标记为转录本一部分的区间重叠交集的次数在通过参考基因组和注释信息对读段进行映射然后再进行表达量估算时候又有两种不同的基因表达量分析方法基因水平分析这种分析方法把一个基因的所有转录本即由基因的不同区域外显子组成的不同版本视为一个整体在这种分析中不区分各个不同的转录本而是将它们合并为一个单一的序列这个序列包括了该基因的所有外显子这个合并的序列被认为是代表性的但它可能并不对应于任何实际存在的单一转录本它更像是一个虚拟的总和对于某些研究目的而言这种简化的方法足够用来获取有意义的数据尤其是当研究的重点是基因整体活动水平时转录本水平分析与基因水平分析不同转录本水平分析关注于每个单独的转录本这种方法不会将不同的转录本合并而是独立地测量每一个转录本的表达量这允许研究者观察到基因内部异构体即同一基因的不同转录本版本的丰度变化这些变化可能对细胞功能和调控至关重要根据转录组进行分类根据转录组进行量化时参考数据将是每个转录本的文件分类将把每个读取分配给一个转录本从而解决过程中的歧义每个转录本的量化矩阵将表示读取被归类为该特定转录本的次数目前主流的能够实现基于分类的工作的软件有我们也能从参考基因组及注释文件中使用工具生成转录组的文件参考基因组通常情况下当我们下载一个参考基因组后我们需要知道一些基本信息比如基因组的大小基因组的组成基因组的注释信息等这些信息对于我们后续的分析非常重要参考基因组大小输出有个序列染色体特定基因组序列标识符总长度为最短序列为最长序列为平均序列长度为有多少个染色体为例统计染色体数结果为每个染色体的长度可通过命令获取每个染色体的长度输出第一行染色体名称第二行染色体长度表示这个染色体的总长度第三行染色体的偏移量这是染色体序列在文件中的起始字节位置这意味着如果你想定位到第号染色体的序列开头你需要从文件的开头跳过个字节第四行每行序列的长度这表示在文件中每行表示的序列的字符数是这不包括行末的换行符第五行每行的长度这个数字比序列的长度多说明每行的结束包括了一个换行符通常是所以每行的实际长度是个字符参考基因组注释文件参考基因组注释文件参考基因组注释遗传标记类型查看参考基因组常见的注释文件格式有和两者都是文本格式的文件这些文件包括了基因组上各种遗传标记的详细信息在分析中基因和转录本的注释通常是最核心的信息因为它们直接关系到后续分析中的表达量计算和差异表达分析基因级别的注释可以提供关于哪些基因在样本中被表达的信息而转录本级别的注释则可以提供更细致的信息比如不同的剪接变体然而是否只需要基因和转录本的遗传标记信息取决于分析的具体目的基因表达定量如果只需要进行基因水平的表达量分析那么基因的注释信息可能就足够了转录本表达定量和剪接分析如果要进行转录本水平的表达分析或者分析剪接变异就必须要有转录本以及外显子的详细注释信息蛋白质编码分析如果分析涉及到蛋白质编码区域例如寻找可能影响蛋白质结构的突变那么编码序列注释就很重要区域分析对于研究转录后调控比如的靶向等和的信息会很关键调控元件分析如果研究的重点是基因的调控机制那么启动子增强子等调控元件的注释就显得非常重要编码序列是指基因组中编码蛋白质的部分这个数字告诉我们在注释文件中有这么多的编码序列条目一种被称为第个氨基酸的罕见氨基酸它直接参与蛋白质的合成这个数字表明在注释文件中有个编码的位点外显子是成熟中保留下来的那部分序列可以转录并翻译成蛋白质这个数字表示注释文件中有这么多外显子非翻译区是位于起始密码子之前的序列部分它不编码蛋白质但有助于调控翻译过程这个数字表示文件中有这么多的条目基因是遗传的基本物理和功能单位通常包含编码一个或多个蛋白质的序列这个数字表示文件中有这么多基因起始密码子是一段特定的核苷酸序列它在上标志着蛋白质合成的开始这个数字显示了有这么多的起始密码子被注释终止密码子是信号蛋白质合成结束的三个核苷酸序列之一这个数字表示有这么多的终止密码子非翻译区是位于终止密码子之后的序列部分它同样不编码蛋白质但含有调控稳定性和翻译效率的重要元素这个数字表示文件中有这么多的条目转录物是指序列通过转录过程生成的副本在这里特指那些最终会被翻译成蛋白质或者具有其他功能的分子这个数字表明有这么多的转录物记录在注释文件中了解数据通常情况下数据是由文件组成文件包含了测序仪输出的原始测序数据每个文件包含了大量的每个都是一个碱基序列同时还包含了测序质量信息我们可以通过查看统计文件来获取数据的一些信息统计文件输出比对构建索引当我们将数据和参考基因组进行比对时首先需要准备参考基因组的索引文件不同的参考基因组比对工具需要不同的索引文件比如需要的索引文件是而需要的索引文件是以为例我们可以通过以下命令构建索引文件比对比对比对率双端比对信息解读总共有个读段所有这些读段都是成对的意味着每个读段都有相应的配对读段有对读段占没有一起成功比对到参考序列上一致比对意味着两个读段以预期的相对方向和距离成功地比对到参考序列的同一个区域有对读段占一致地比对到参考序列的唯一位置一次有对读段占一致地比对到参考序列的多个位置接下来的部分提供了关于没有一致比对的读段的更多细节在没有一致比对的对读段中有对占以不一致的方式比对了一次不一致比对意味着这些读段没有按照预期的相对方向或距离比对到参考序列的同一个区域但仍然在某处找到了比对位置在这对既没有一致比对也没有不一致比对的读段中总共有个单独的读段其中个占没有比对到参考序列有个读段占比对到了参考序列的唯一位置一次有个读段占比对到了参考序列的多个位置最后一行提供了总体的比对率总体来说的读段至少比对到了参考序列的一个位置这个百分比是基于所有读段的包括一致比对的不一致比对的以及那些只有一个读段比对到参考序列的配对读段单端比对信息解读表示总共有个读段被处理读段是指从测序实验得到的短序列片段所有的读段都是单端读段意味着这些读段不是成对的在配对端测序中你通常会得到两个读段即一对它们是从同一个片段的两端测序得到的单端读段意味着每个片段只测序了一端在这些读段中有个占总数的没有比对到参考序列上个读段占总数的精确地比对到参考序列上一次这意味着这些读段在参考序列中有一个唯一的匹配位置有个读段占总数的比对到参考序列上多于一次这表明这些读段可能来自基因组或转录组中重复区域因此无法确定它们的精确位置文件转换通常情况下我们可以将文件直接在等可视化软件中查看但是有时候加载时间很长当我们只是关注覆盖率时我们可以将文件转换为格式这样加载速度会更快在这个过程中我们需要将文件转换为文件然后再转换为文件计数参数标准化计数矩阵得到比对文件后我们需要对比对文件进行计数得到每个基因的表达量这里我们可以使用工具是软件包的一部分它可以用于计算基因或转录本的表达量此时命令参数是必不可少的我们要确保命令按照文件中列出的顺序执行每当顺序至关重要时您必须使用否则顺序就无法保证当是双端测序的时候我们可以设置参数这样可以计算每对的计数非必选当我们使用时我们会做出许多决定除了前面提到的命令外我们还可以使用其他参数来调整计数的方式下面是一些常用的参数生成的文件包含了每个基因的表达量信息可使用自定义脚本来转化为更简单直接的数据差异表达分析差异表达分析得到计数矩阵后我们就可以进行差异表达分析了常用的差异表达分析工具有等基于线性模型通过使用贝叶斯方法估计每个基因的差异方差它使用经验贝叶斯方法来将信息从所有基因中借用特别是在样本较少时提高估计的稳定性基于负二项分布模型它使用贝叶斯方法通过适应组内变异估计提高估计的稳定性考虑了基因的丰度和变异性使其更适用于数据基于负二项分布的模型它通过使用贝叶斯方法来考虑样本间的差异以及基因表达的离散性这里我们使用进行差异表达分析我们需要用到两个文件一个是且是我们在上一步骤简化过的另一个是它包含了实验设计信息一列是一列是是我们要比较的分组信息分析所遇问题比对问题编码序列的含量指基因序列中鸟嘌呤和胞嘧啶的比例通常高于整个基因组的平均含量这是因为编码序列往往需要维持一定的稳定性对之间的氢键比对更多增加了双螺旋的稳定性在基因组测序和比对分析中如果一个区域的含量远离平均水平不管是偏高还是偏低都可能导致测序和分析上的偏差例如高含量的区域可能因为扩增偏好性或测序平台的偏好性而过度表示此外极端的含量可能影响测序读段的比对因为这样的区域在基因组中可能不是唯一的或者比对算法在处理这些区域时可能不够准确查看序列含量查看序列含量特别的当我们发现数据比对出现长内含子区域时可能就是由于参考序列中较长的仅有组成造成的在中我们可以使用参数来调整最大内含子长度以提高比对正确率经典的内含子长度为我们可以将这个数值设置为计数问题在使用时默认使用进行计数通常有两种策略一种是在基因水平分析一种是在转录本水平分析在基因水平分析时外显子只会被计数一次而在转录本水平时外显子可能出现在多个转录本中如果简单的每个外显子计数一次可能会导致计数的重复导致从贡献度被过度估计为了解决这个问题需要一个过程来确定每个外显子对每个转录本的贡献比例这个过程称为解卷积解卷积的目的是分配每个外显子的表达量确保即使一个外显子在多个转录本中出现它的总贡献也只被计算一次解卷积通常涉及到数学和统计方法可以基于转录本的预期丰度或其他生物信息学数据来估计每个外显子的相对贡献一般情况下建议使用进行基因水平分析使用基于分类伪比对的方法如和等工具进行转录水平分析链特异性问题文库类型的选择决定了测序的读取方向和链的测序顺序这意味着来自不同文库类型的读取可能存在显著差异有关文库类型的信息对于将读取组装成转录组或映射到参考组装非常有用这是因为文库类型可以帮助通过读取的相对方向和从哪条链测序来帮助辨别转录组中较短的模糊读取属于哪里不幸的是有关所用文库类型的信息未包含在测序输出文件中并且可能在数据组装之前丢失即使使用来自公共存储库的数据也不能保证文库类型信息是正确的或根本不存在在成对端测序中每对测序读取包括一个正向和一个反向组分这些读取对应于片段的两端通常用于提高测序的覆盖率和准确性然而当数据是成对的且有方向性时理解每个读取是如何对应于参考基因组或转录本的变得更加复杂在有方向性的成对端对齐中通常第一对中的读取会与反义链对齐而第二对中的读取则会与正义链对齐例如如果来自于反向链上的一个转录本那么第一对中的读取会对齐到正向链上并且会在参考序列上向左移动第二对中的读取会对齐到反向链并且相对于片段会向右移动虽然这听起来简单实际上却不是这样在测序数据的方向性问题上存在很多混淆和误解以至于有专门的软件工具例如来帮助研究人员弄清楚他们的数据是否有方向性以及是哪种类型的方向性在一些软件中有参数可以用来指定来链的方向转录完整性分析流程比对和组装定量和标准化差异分析差异分析基因表达量数据类型基因表达量数据转换原始数据基因长度信息外显子长度之和如按照基因计算计算还需提供基因的分类信息',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2024-06-10 23:10:01',
  postMainColor: '',
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
      win.saveToLocal = {
        set: (key, value, ttl) => {
          if (ttl === 0) return
          const now = Date.now()
          const expiry = now + ttl * 86400000
          const item = {
            value,
            expiry
          }
          localStorage.setItem(key, JSON.stringify(item))
        },
      
        get: key => {
          const itemStr = localStorage.getItem(key)
      
          if (!itemStr) {
            return undefined
          }
          const item = JSON.parse(itemStr)
          const now = Date.now()
      
          if (now > item.expiry) {
            localStorage.removeItem(key)
            return undefined
          }
          return item.value
        }
      }
    
      win.getScript = (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        script.onerror = reject
        script.onload = script.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          script.onload = script.onreadystatechange = null
          resolve()
        }

        Object.keys(attr).forEach(key => {
          script.setAttribute(key, attr[key])
        })

        document.head.appendChild(script)
      })
    
      win.getCSS = (url, id = false) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onerror = reject
        link.onload = link.onreadystatechange = function() {
          const loadState = this.readyState
          if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
          link.onload = link.onreadystatechange = null
          resolve()
        }
        document.head.appendChild(link)
      })
    
      win.activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#18171d')
        }
      }
      win.activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#f7f9fe')
        }
      }
      const t = saveToLocal.get('theme')
    
          const isDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches
          const isLightMode = window.matchMedia('(prefers-color-scheme: light)').matches
          const isNotSpecified = window.matchMedia('(prefers-color-scheme: no-preference)').matches
          const hasNoSupport = !isDarkMode && !isLightMode && !isNotSpecified

          if (t === undefined) {
            if (isLightMode) activateLightMode()
            else if (isDarkMode) activateDarkMode()
            else if (isNotSpecified || hasNoSupport) {
              const now = new Date()
              const hour = now.getHours()
              const isNight = hour <= 6 || hour >= 18
              isNight ? activateDarkMode() : activateLightMode()
            }
            window.matchMedia('(prefers-color-scheme: dark)').addListener(e => {
              if (saveToLocal.get('theme') === undefined) {
                e.matches ? activateDarkMode() : activateLightMode()
              }
            })
          } else if (t === 'light') activateLightMode()
          else activateDarkMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
      const detectApple = () => {
        if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
          document.documentElement.classList.add('apple')
        }
      }
      detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"><link rel="alternate" href="/atom.xml" title="海底捞里没有鱼" type="application/atom+xml">
</head><body data-type="anzhiyu"><div id="web_bg"></div><div id="an_music_bg"></div><div id="loading-box" onclick="document.getElementById(&quot;loading-box&quot;).classList.add(&quot;loaded&quot;)"><div class="loading-bg"><img class="loading-img nolazyload" alt="加载头像" src="https://bu.dusays.com/2023/06/10/648499e1299d0.jpg"/><div class="loading-image-dot"></div></div></div><script>const preloader = {
  endLoading: () => {
    document.getElementById('loading-box').classList.add("loaded");
  },
  initLoading: () => {
    document.getElementById('loading-box').classList.remove("loaded")
  }
}
window.addEventListener('load',()=> { preloader.endLoading() })
setTimeout(function(){preloader.endLoading();},10000)

if (true) {
  document.addEventListener('pjax:send', () => { preloader.initLoading() })
  document.addEventListener('pjax:complete', () => { preloader.endLoading() })
}</script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.0.0/progress_bar/progress_bar.css"/><script async="async" src="https://cdn.cbd.int/pace-js@1.2.4/pace.min.js" data-pace-options="{ &quot;restartOnRequestAfter&quot;:false,&quot;eventLag&quot;:false}"></script><div class="post" id="body-wrap"><header class="post-bg" id="page-header"><nav id="nav"><div id="nav-group"><span id="blog_name"><div class="back-home-button"><i class="anzhiyufont anzhiyu-icon-grip-vertical"></i><div class="back-menu-list-groups"><div class="back-menu-list-group"><div class="back-menu-list-title">网页</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://smileszh.cn" title="博客首页"><img class="back-menu-item-icon" src="/img/favicon.ico" alt="博客首页"/><span class="back-menu-item-text">博客首页</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">项目</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://github.com/smileszh/smileszh.github.io" title="Github"><img class="back-menu-item-icon" src="https://bu.dusays.com/2023/06/07/648036b97aaba.png" alt="Github"/><span class="back-menu-item-text">Github</span></a><a class="back-menu-item" target="_blank" rel="noopener" href="http://127.0.0.1:5244/" title="alist"><img class="back-menu-item-icon" src="https://bu.dusays.com/2023/07/12/64adad75d8996.png" alt="alist"/><span class="back-menu-item-text">alist</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">图床</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://7bu.top/" title="去不图床"><img class="back-menu-item-icon" src="https://bu.dusays.com/2023/06/07/6480380fd3f0b.png" alt="去不图床"/><span class="back-menu-item-text">去不图床</span></a><a class="back-menu-item" target="_blank" rel="noopener" href="https://image.anheyu.com/" title="安和鱼图床"><img class="back-menu-item-icon" src="https://bu.dusays.com/2023/06/07/6480380fd3f0b.png" alt="安和鱼图床"/><span class="back-menu-item-text">安和鱼图床</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">云盘</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://yun.console.xlj0.com/login?redirect=%2Flogin" title="Tooc公益网盘"><img class="back-menu-item-icon" src="https://bu.dusays.com/2023/06/20/6490a64ff04d3.png" alt="Tooc公益网盘"/><span class="back-menu-item-text">Tooc公益网盘</span></a></div></div></div></div><a id="site-name" href="/" accesskey="h"><div class="title">海底捞里没有鱼</div><i class="anzhiyufont anzhiyu-icon-house-chimney"></i></a></span><div class="mask-name-container"><div id="name-container"><a id="page-name" href="javascript:anzhiyu.scrollToDest(0, 500)">PAGE_NAME</a></div></div><div id="menus"><div class="menus_items"><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 文章</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/archives/"><i class="anzhiyufont anzhiyu-icon-box-archive faa-tada" style="font-size: 0.9em;"></i><span> 隧道</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/categories/"><i class="anzhiyufont anzhiyu-icon-shapes faa-tada" style="font-size: 0.9em;"></i><span> 分类</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags faa-tada" style="font-size: 0.9em;"></i><span> 标签</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 友链</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/link/"><i class="anzhiyufont anzhiyu-icon-link faa-tada" style="font-size: 0.9em;"></i><span> 友人帐</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/fcircle/"><i class="anzhiyufont anzhiyu-icon-artstation faa-tada" style="font-size: 0.9em;"></i><span> 朋友圈</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/comments/"><i class="anzhiyufont anzhiyu-icon-envelope faa-tada" style="font-size: 0.9em;"></i><span> 留言板</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 生活</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/music/"><i class="anzhiyufont anzhiyu-icon-music faa-tada" style="font-size: 0.9em;"></i><span> 音乐馆</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/bangumis/"><i class="anzhiyufont anzhiyu-icon-bilibili faa-tada" style="font-size: 0.9em;"></i><span> 追番页</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/album/"><i class="anzhiyufont anzhiyu-icon-images faa-tada" style="font-size: 0.9em;"></i><span> 相册集</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 我的</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/about/"><i class="anzhiyufont anzhiyu-icon-paper-plane faa-tada" style="font-size: 0.9em;"></i><span> 关于我们</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/essay/"><i class="anzhiyufont anzhiyu-icon-lightbulb faa-tada" style="font-size: 0.9em;"></i><span> 闲言碎语</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/equipment/"><i class="anzhiyufont anzhiyu-icon-dice-d20 faa-tada" style="font-size: 0.9em;"></i><span> 我的装备</span></a></li></ul></div></div></div><div id="nav-right"><div class="nav-button only-home" id="travellings_button" title="随机前往一个开往项目网站"><a class="site-page" onclick="anzhiyu.totraveling()" title="随机前往一个开往项目网站" href="javascript:void(0);" rel="external nofollow" data-pjax-state="external"><i class="anzhiyufont anzhiyu-icon-train"></i></a></div><div class="nav-button" id="randomPost_button"><a class="site-page" onclick="toRandomPost()" title="随机前往一个文章" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-dice"></i></a></div><div class="nav-button" id="search-button"><a class="site-page social-icon search" href="javascript:void(0);" title="搜索🔍" accesskey="s"><i class="anzhiyufont anzhiyu-icon-magnifying-glass"></i><span> 搜索</span></a></div><input id="center-console" type="checkbox"/><label class="widget" for="center-console" title="中控台" onclick="anzhiyu.switchConsole();"><i class="left"></i><i class="widget center"></i><i class="widget right"></i></label><div id="console"><div class="console-card-group-reward"><ul class="reward-all console-card"><li class="reward-item"><a href="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-weichat.png" target="_blank"><img class="post-qr-code-img" alt="微信" src="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-weichat.png"/></a><div class="post-qr-code-desc">微信</div></li><li class="reward-item"><a href="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-alipay.png" target="_blank"><img class="post-qr-code-img" alt="支付宝" src="https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/qrcode-alipay.png"/></a><div class="post-qr-code-desc">支付宝</div></li></ul></div><div class="console-card-group"><div class="console-card-group-left"><div class="console-card" id="card-newest-comments"><div class="card-content"><div class="author-content-item-tips">互动</div><span class="author-content-item-title"> 最新评论</span></div><div class="aside-list"><span>正在加载中...</span></div></div></div><div class="console-card-group-right"><div class="console-card tags"><div class="card-content"><div class="author-content-item-tips">兴趣点</div><span class="author-content-item-title">寻找你感兴趣的领域</span><div class="card-tags"><div class="item-headline"></div><div class="card-tag-cloud"><a href="/tags/GO-KEGG/" style="font-size: 1.05rem;">GO/KEGG<sup>2</sup></a><a href="/tags/Mac/" style="font-size: 1.05rem;">Mac<sup>1</sup></a><a href="/tags/R-%E8%AF%AD%E8%A8%80/" style="font-size: 1.05rem;">R 语言<sup>1</sup></a><a href="/tags/R%E5%8C%85/" style="font-size: 1.05rem;">R包<sup>4</sup></a><a href="/tags/R%E7%BB%98%E5%9B%BE/" style="font-size: 1.05rem;">R绘图<sup>2</sup></a><a href="/tags/R%E8%AF%AD%E8%A8%80/" style="font-size: 1.05rem;">R语言<sup>9</sup></a><a href="/tags/Terminal/" style="font-size: 1.05rem;">Terminal<sup>1</sup></a><a href="/tags/conda/" style="font-size: 1.05rem;">conda<sup>1</sup></a><a href="/tags/tidyverse/" style="font-size: 1.05rem;">tidyverse<sup>1</sup></a><a href="/tags/%E5%88%86%E5%AD%90%E4%BA%92%E4%BD%9C/" style="font-size: 1.05rem;">分子互作<sup>1</sup></a><a href="/tags/%E5%9F%BA%E5%9B%A0ID/" style="font-size: 1.05rem;">基因ID<sup>2</sup></a><a href="/tags/%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A/" style="font-size: 1.05rem;">基因注释<sup>2</sup></a><a href="/tags/%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" style="font-size: 1.05rem;">富集分析<sup>1</sup></a><a href="/tags/%E5%B7%AE%E5%BC%82%E8%A1%A8%E6%B3%95%E5%88%86%E6%9E%90/" style="font-size: 1.05rem;">差异表法分析<sup>1</sup></a><a href="/tags/%E5%BC%95%E7%89%A9%E8%AE%BE%E8%AE%A1/" style="font-size: 1.05rem;">引物设计<sup>1</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" style="font-size: 1.05rem;">数据分析<sup>1</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 1.05rem;">数据库<sup>3</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F/" style="font-size: 1.05rem;">数据格式<sup>1</sup></a><a href="/tags/%E7%94%9F%E7%89%A9%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 1.05rem;">生物数据库<sup>1</sup></a><a href="/tags/%E7%9F%A5%E8%AF%86%E7%82%B9/" style="font-size: 1.05rem;">知识点<sup>1</sup></a><a href="/tags/%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90/" style="font-size: 1.05rem;">聚类分析<sup>1</sup></a><a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 1.05rem;">转录组<sup>1</sup></a><a href="/tags/%E8%BD%AF%E4%BB%B6%E7%AE%A1%E7%90%86/" style="font-size: 1.05rem;">软件管理<sup>1</sup></a></div></div><hr/></div></div><div class="console-card history"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-box-archiv"></i><span>文章</span></div><div class="card-archives"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-archive"></i><span>归档</span></div><ul class="card-archive-list"><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/06/"><span class="card-archive-list-date">六月 2024</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">1</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/05/"><span class="card-archive-list-date">五月 2024</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">2</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2024/04/"><span class="card-archive-list-date">四月 2024</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">1</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2023/08/"><span class="card-archive-list-date">八月 2023</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">1</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2023/07/"><span class="card-archive-list-date">七月 2023</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">3</span><span>篇</span></div></a></li><li class="card-archive-list-item"><a class="card-archive-list-link" href="/archives/2023/06/"><span class="card-archive-list-date">六月 2023</span><div class="card-archive-list-count-group"><span class="card-archive-list-count">15</span><span>篇</span></div></a></li></ul></div><hr/></div></div></div><div class="button-group"><div class="console-btn-item"><a class="darkmode_switchbutton" title="显示模式切换" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-moon"></i></a></div><div class="console-btn-item" id="consoleHideAside" onclick="anzhiyu.hideAsideBtn()" title="边栏显示控制"><a class="asideSwitch"><i class="anzhiyufont anzhiyu-icon-arrows-left-right"></i></a></div><div class="console-btn-item on" id="consoleCommentBarrage" onclick="anzhiyu.switchCommentBarrage()" title="热评开关"><a class="commentBarrage"><i class="anzhiyufont anzhiyu-icon-message"></i></a></div><div class="console-btn-item" id="consoleMusic" onclick="anzhiyu.musicToggle()" title="音乐开关"><a class="music-switch"><i class="anzhiyufont anzhiyu-icon-music"></i></a></div></div><div class="console-mask" onclick="anzhiyu.hideConsole()" href="javascript:void(0);"></div></div><div class="nav-button" id="nav-totop"><a class="totopbtn" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i><span id="percent" onclick="anzhiyu.scrollToDest(0,500)">0</span></a></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);" title="切换"><i class="anzhiyufont anzhiyu-icon-bars"></i></a></div></div></div></nav><div id="post-info"><div id="post-firstinfo"><div class="meta-firstline"><a class="post-meta-original">原创</a><span class="post-meta-categories"><span class="post-meta-separator"></span><i class="anzhiyufont anzhiyu-icon-inbox post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" itemprop="url">生物信息学</a></span><span class="article-meta tags"><a class="article-meta__tags" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" tabindex="-1" itemprop="url"> <span> <i class="anzhiyufont anzhiyu-icon-hashtag"></i>转录组</span></a></span></div></div><h1 class="post-title" itemprop="name headline">RNA-Seq</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="anzhiyufont anzhiyu-icon-calendar-days post-meta-icon"></i><span class="post-meta-label">发表于</span><time class="post-meta-date-created" itemprop="dateCreated datePublished" datetime="2023-06-29T08:00:00.000Z" title="发表于 2023-06-29 16:00:00">2023-06-29</time><span class="post-meta-separator"></span><i class="anzhiyufont anzhiyu-icon-history post-meta-icon"></i><span class="post-meta-label">更新于</span><time class="post-meta-date-updated" itemprop="dateCreated datePublished" datetime="2024-06-10T15:10:01.527Z" title="更新于 2024-06-10 23:10:01">2024-06-10</time></span></div><div class="meta-secondline"><span class="post-meta-separator"></span><span class="post-meta-pv-cv" id="" data-flag-title="RNA-Seq"><i class="anzhiyufont anzhiyu-icon-fw-eye post-meta-icon"></i><span class="post-meta-label" title="阅读量">阅读量:</span><span id="busuanzi_value_page_pv"><i class="anzhiyufont anzhiyu-icon-spinner anzhiyu-spin"></i></span></span><span class="post-meta-separator">       </span><span class="post-meta-position" title="作者IP属地为宁夏"><i class="anzhiyufont anzhiyu-icon-location-dot"></i>宁夏</span></div></div></div><section class="main-hero-waves-area waves-area"><svg class="waves-svg" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" viewBox="0 24 150 28" preserveAspectRatio="none" shape-rendering="auto"><defs><path id="gentle-wave" d="M -160 44 c 30 0 58 -18 88 -18 s 58 18 88 18 s 58 -18 88 -18 s 58 18 88 18 v 44 h -352 Z"></path></defs><g class="parallax"><use href="#gentle-wave" x="48" y="0"></use><use href="#gentle-wave" x="48" y="3"></use><use href="#gentle-wave" x="48" y="5"></use><use href="#gentle-wave" x="48" y="7"></use></g></svg></section><div id="post-top-cover"><img class="nolazyload" id="post-top-bg" src="https://bu.dusays.com/2023/07/03/64a283d519f00.png"></div></header><main id="blog-container"><div class="layout" id="content-inner"><div id="post"><article class="post-content" id="article-container" itemscope itemtype="https://smileszh.github.io/posts/7acd0447.html"><header><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/" itemprop="url">生物信息学</a><a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" tabindex="-1" itemprop="url">转录组</a><h1 id="CrawlerTitle" itemprop="name headline">RNA-Seq</h1><span itemprop="author" itemscope itemtype="http://schema.org/Person">泡泡</span><time itemprop="dateCreated datePublished" datetime="2023-06-29T08:00:00.000Z" title="发表于 2023-06-29 16:00:00">2023-06-29</time><time itemprop="dateCreated datePublished" datetime="2024-06-10T15:10:01.527Z" title="更新于 2024-06-10 23:10:01">2024-06-10</time></header><h1 id="RNA-Seq指南"><a href="#RNA-Seq指南" class="headerlink" title="RNA-Seq指南"></a>RNA-Seq指南</h1><h2 id="RNA-Seq参考"><a href="#RNA-Seq参考" class="headerlink" title="RNA-Seq参考"></a>RNA-Seq参考</h2><h3 id="根据基因组进行量化"><a href="#根据基因组进行量化" class="headerlink" title="根据基因组进行量化"></a>根据基因组进行量化</h3><p>RNA-Seq根据基因组进行量化时，参考数据是基因组序列，基因组注释文件是基因组的注释信息，包括基因的位置、外显子的位置、基因的功能等信息。该方法将对齐文件与注释相交叉，以产生“计数”，然后过滤这些计数以保留统计上显著的结果。定量矩阵代表对齐与标记为转录本一部分的区间重叠（交集）的次数。</p>
<div class="img-wrap"><div class="img-bg"><img class="img" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/06/07/6662c38b4ccfa.png"/></div></div>
<p>在通过参考基因组和注释信息对读段进行映射，然后再进行表达量估算时候，又有两种不同的基因表达量分析方法。</p>
<ol>
<li><strong>基因水平分析</strong>：</li>
</ol>
<ul>
<li>这种分析方法把一个基因的所有转录本（即由基因的不同区域（外显子）组成的不同版本）视为一个整体。</li>
<li>在这种分析中，不区分各个不同的转录本，而是将它们合并为一个单一的序列，这个序列包括了该基因的所有外显子。</li>
<li>这个合并的序列被认为是“代表性”的，但它可能并不对应于任何实际存在的单一转录本，它更像是一个虚拟的总和。</li>
<li>对于某些研究目的而言，这种简化的方法足够用来获取有意义的数据，尤其是当研究的重点是基因整体活动水平时。</li>
</ul>
<ol>
<li><strong>转录本水平分析</strong>：</li>
</ol>
<ul>
<li>与基因水平分析不同，转录本水平分析关注于每个单独的转录本。</li>
<li>这种方法不会将不同的转录本合并，而是独立地测量每一个转录本的表达量。</li>
<li>这允许研究者观察到基因内部异构体（即同一基因的不同转录本版本）的丰度变化，这些变化可能对细胞功能和调控至关重要。</li>
</ul>
<h3 id="根据转录组进行分类"><a href="#根据转录组进行分类" class="headerlink" title="根据转录组进行分类"></a>根据转录组进行分类</h3><p>RNA-Seq根据转录组进行量化时，参考数据将是每个转录本的 FASTA 文件。分类将把每个读取分配给一个转录本，从而解决过程中的歧义。每个转录本的量化矩阵将表示读取被归类为该特定转录本的次数。</p>
<div class="img-wrap"><div class="img-bg"><img class="img" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/06/07/6662c3d560ec8.png"/></div></div>
<p>目前主流的能够实现基于分类的RNA-Seq工作的软件有：</p>
<ul>
<li><strong>Salmon</strong></li>
<li><strong>Kallisto</strong></li>
</ul>
<p>我们也能从参考基因组及注释文件中使用<code>gffread</code>工具生成转录组的FASTA文件。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">conda install gffread</span><br><span class="line"></span><br><span class="line"><span class="comment"># Extract the transcripts.</span></span><br><span class="line">gffread -w refs/transcripts.fa -g refs/22.fa refs/22.gtf</span><br></pre></td></tr></table></figure></p>
<h2 id="参考基因组"><a href="#参考基因组" class="headerlink" title="参考基因组"></a>参考基因组</h2><p>通常情况下，当我们下载一个参考基因组后，我们需要知道一些基本信息，比如基因组的大小、基因组的组成、基因组的注释信息等。这些信息对于我们后续的分析非常重要。</p>
<h3 id="参考基因组大小"><a href="#参考基因组大小" class="headerlink" title="参考基因组大小"></a>参考基因组大小</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">seqkit stats *.fa</span><br></pre></td></tr></table></figure>
<p>输出：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">file        format  <span class="built_in">type</span>  num_seqs        sum_len  min_len       avg_len      max_len</span><br><span class="line">genomic.fa  FASTA   DNA        194  3,099,750,718      970  15,978,096.5  248,956,422</span><br></pre></td></tr></table></figure><br>有194个序列（染色体+特定基因组序列标识符），总长度为3,099,750,718bp，最短序列为970bp，最长序列为248,956,422bp，平均序列长度为15,978,096.5bp。</p>
<h3 id="有多少个染色体"><a href="#有多少个染色体" class="headerlink" title="有多少个染色体"></a>有多少个染色体</h3><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> *.fa | grep <span class="string">&quot;&gt;&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Ensemble GRCh38w为例，统计染色体数。（结果为22+X+Y+MT=25）</span></span><br><span class="line"><span class="built_in">cat</span> *.fa | grep <span class="string">&quot;&gt;*chromosome*&quot;</span> | <span class="built_in">wc</span> -l</span><br></pre></td></tr></table></figure>
<h3 id="每个染色体的长度"><a href="#每个染色体的长度" class="headerlink" title="每个染色体的长度"></a>每个染色体的长度</h3><p>可通过samtools faidx命令获取每个染色体的长度。<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">samtools faidx *.fa</span><br><span class="line"><span class="built_in">cat</span> *.fa.fai</span><br></pre></td></tr></table></figure><br>输出：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">1       248956422       56      60      61</span><br><span class="line">10      133797422       253105810       60      61</span><br><span class="line">11      135086622       389133248       60      61</span><br><span class="line">12      133275309       526471372       60      61</span><br><span class="line">13      114364328       661967995       60      61</span><br><span class="line">14      107043718       778238454       60      61</span><br><span class="line">15      101991189       887066292       60      61</span><br><span class="line">16      90338345        990757392       60      61</span><br><span class="line">17      83257441        1082601434      60      61</span><br><span class="line">18      80373285        1167246557      60      61</span><br><span class="line">···</span><br><span class="line">···</span><br></pre></td></tr></table></figure></p>
<p><code>第一行</code>：染色体名称。<br><code>第二行</code>：染色体长度，表示这个染色体的总长度。<br><code>第三行</code>：染色体的偏移量，这是染色体序列在FASTA文件中的起始字节位置。这意味着，如果你想定位到第10号染色体的序列开头，你需要从文件的开头跳过253,105,810个字节。<br><code>第四行</code>：每行序列的长度，<strong>60</strong>，这表示在FASTA文件中，每行表示的序列的字符数是60。这不包括行末的换行符。<br><code>第五行</code>：每行的长度，<strong>61</strong>，这个数字比序列的长度多1，说明每行的结束包括了一个换行符（通常是<code>\n</code>），所以每行的实际长度是61个字符。</p>
<h2 id="参考基因组注释文件"><a href="#参考基因组注释文件" class="headerlink" title="参考基因组注释文件"></a>参考基因组注释文件</h2><div class="tabs" id="注释文件"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="注释文件-1">参考基因组注释文件</button><button type="button" class="tab " data-href="注释文件-2">参考基因组注释遗传标记类型查看</button></ul><div class="tab-contents"><div class="tab-item-content active" id="注释文件-1"><p>参考基因组常见的注释文件格式有GTF（Gene Transfer Format）和 GFF（General Feature Format），两者都是文本格式的文件。这些文件包括了基因组上各种遗传标记的详细信息。<br>在RNA-Seq分析中，基因和转录本的注释通常是最核心的信息，因为它们直接关系到后续分析中的表达量计算和差异表达分析。基因级别的注释可以提供关于哪些基因在样本中被表达的信息，而转录本级别的注释则可以提供更细致的信息，比如不同的剪接变体。</p>
<p>然而，是否只需要基因和转录本的遗传标记信息取决于分析的具体目的：</p>
<ol>
<li><p><strong>基因表达定量</strong>: 如果只需要进行基因水平的表达量分析，那么基因的注释信息可能就足够了。</p>
</li>
<li><p><strong>转录本表达定量和剪接分析</strong>: 如果要进行转录本水平的表达分析或者分析剪接变异，就必须要有转录本以及外显子的详细注释信息。</p>
</li>
<li><p><strong>蛋白质编码分析</strong>: 如果分析涉及到蛋白质编码区域，例如寻找可能影响蛋白质结构的突变，那么编码序列（CDS）注释就很重要。</p>
</li>
<li><p><strong>UTR区域分析</strong>: 对于研究转录后调控，比如microRNA的靶向等，5’ UTR和3’ UTR的信息会很关键。</p>
</li>
<li><p><strong>调控元件分析</strong>: 如果研究的重点是基因的调控机制，那么启动子、增强子等调控元件的注释就显得非常重要。</p>
</li>
</ol></div><div class="tab-item-content" id="注释文件-2"><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">cat</span> *.gtf | grep -v <span class="string">&quot;#&quot;</span> | <span class="built_in">cut</span> -f3 | <span class="built_in">sort</span> | <span class="built_in">uniq</span> -c</span><br><span class="line"></span><br><span class="line">746300 CDS</span><br><span class="line">    120 Selenocysteine</span><br><span class="line">1261907 exon</span><br><span class="line"> 149893 five_prime_utr</span><br><span class="line">  58676 gene</span><br><span class="line">  86423 start_codon</span><br><span class="line">  78536 stop_codon</span><br><span class="line"> 148456 three_prime_utr</span><br><span class="line"> 206534 transcript</span><br></pre></td></tr></table></figure>
<ol>
<li><p><strong>CDS (Coding Sequence)</strong>: 编码序列是指基因组中编码蛋白质的部分，这个数字（746300）告诉我们在注释文件中有这么多的编码序列条目。</p>
</li>
<li><p><strong>Selenocysteine</strong>: 一种被称为“第21个氨基酸”的罕见氨基酸，它直接参与蛋白质的合成。这个数字（120）表明在注释文件中有120个编码selenocysteine的位点。</p>
</li>
<li><p><strong>Exon</strong>: 外显子是成熟mRNA中保留下来的那部分DNA序列，可以转录并翻译成蛋白质。这个数字（1261907）表示注释文件中有这么多外显子。</p>
</li>
<li><p><strong>Five_prime_utr (5’ UTR)</strong>: 5’非翻译区（Untranslated Region）是位于起始密码子之前的mRNA序列部分，它不编码蛋白质，但有助于调控翻译过程。这个数字（149893）表示文件中有这么多的5’ UTR条目。</p>
</li>
<li><p><strong>Gene</strong>: 基因是遗传的基本物理和功能单位，通常包含编码一个或多个蛋白质的序列。这个数字（58676）表示文件中有这么多基因。</p>
</li>
<li><p><strong>Start_codon</strong>: 起始密码子是一段特定的核苷酸序列，它在mRNA上标志着蛋白质合成的开始。这个数字（86423）显示了有这么多的起始密码子被注释。</p>
</li>
<li><p><strong>Stop_codon</strong>: 终止密码子是信号蛋白质合成结束的三个核苷酸序列之一。这个数字（78536）表示有这么多的终止密码子。</p>
</li>
<li><p><strong>Three_prime_utr (3’ UTR)</strong>: 3’非翻译区（Untranslated Region）是位于终止密码子之后的mRNA序列部分，它同样不编码蛋白质，但含有调控mRNA稳定性和翻译效率的重要元素。这个数字（148456）表示文件中有这么多的3’ UTR条目。</p>
</li>
<li><p><strong>Transcript</strong>: 转录物是指DNA序列通过转录过程生成的RNA副本，在这里特指那些最终会被翻译成蛋白质或者具有其他功能的RNA分子。这个数字（206534）表明有这么多的转录物记录在注释文件中。</p>
</li>
</ol></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div>
<h2 id="了解数据"><a href="#了解数据" class="headerlink" title="了解数据"></a>了解数据</h2><p>通常情况下RNA-Seq数据是由fastq文件组成，fastq文件包含了测序仪输出的原始测序数据，每个fastq文件包含了大量的reads，每个read都是一个碱基序列，同时还包含了测序质量信息。我们可以通过查看统计fastq文件来获取数据的一些信息。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 统计fastq文件</span></span><br><span class="line">seqkit stats *.fq</span><br></pre></td></tr></table></figure>
<p>输出：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">file               format  <span class="built_in">type</span>    num_seqs        sum_len  min_len  avg_len  max_len</span><br><span class="line">chip_HA_1_1.fq     FASTQ   DNA   40,478,593  2,023,929,650       50       50       50</span><br><span class="line">chip_HA_2_1.fq     FASTQ   DNA   37,248,768  1,862,438,400       50       50       50</span><br><span class="line">chip_HA_3_1.fq     FASTQ   DNA   28,160,024  1,408,001,200       50       50       50</span><br><span class="line">chip_input_1_1.fq  FASTQ   DNA   42,065,898  2,103,294,900       50       50       50</span><br><span class="line">chip_plau_1_1.fq   FASTQ   DNA   31,915,339  1,595,766,950       50       50       50</span><br><span class="line">chip_plau_2_1.fq   FASTQ   DNA   43,102,269  2,155,113,450       50       50       50</span><br><span class="line">chip_plau_3_1.fq   FASTQ   DNA   38,000,000  1,900,000,000       50       50       50</span><br></pre></td></tr></table></figure></p>
<h2 id="比对"><a href="#比对" class="headerlink" title="比对"></a>比对</h2><h3 id="构建索引"><a href="#构建索引" class="headerlink" title="构建索引"></a>构建索引</h3><p>当我们将数据和参考基因组进行比对时，首先需要准备参考基因组的索引文件，不同的参考基因组比对工具需要不同的索引文件，比如STAR需要的索引文件是STAR_index，而HISAT2需要的索引文件是HISAT2_index。<br>以Hisat2为例，我们可以通过以下命令构建索引文件：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># The reference genome.</span></span><br><span class="line">REF=refs/*.fa</span><br><span class="line">GTF=refs/*.gtf</span><br><span class="line"></span><br><span class="line"><span class="comment"># Build the genome index.</span></span><br><span class="line">IDX=refs/genome</span><br><span class="line"></span><br><span class="line">hisat2_extract_splice_sites.py <span class="variable">$&#123;GTF&#125;</span> &gt; <span class="variable">$&#123;IDX&#125;</span>.ss</span><br><span class="line">hisat2_extract_exons.py <span class="variable">$&#123;GTF&#125;</span> &gt; <span class="variable">$&#123;IDX&#125;</span>.exon</span><br><span class="line"></span><br><span class="line">hisat2-build -p 16 --exon <span class="variable">$&#123;IDX&#125;</span>.exon --ss <span class="variable">$&#123;IDX&#125;</span>.ss <span class="variable">$&#123;REF&#125;</span> <span class="variable">$&#123;REF&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Index the reference genome with samtools.</span></span><br><span class="line">samtools faidx <span class="variable">$&#123;REF&#125;</span></span><br></pre></td></tr></table></figure></p>
<h3 id="比对-1"><a href="#比对-1" class="headerlink" title="比对"></a>比对</h3><div class="tabs" id="比对"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="比对-1">HISAT2比对</button><button type="button" class="tab " data-href="比对-2">HISAT比对率</button></ul><div class="tab-contents"><div class="tab-item-content active" id="比对-1"><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Create the include sample list file.</span></span><br><span class="line"><span class="built_in">ls</span> reads | <span class="built_in">cut</span> -d <span class="string">&#x27;_&#x27;</span> -f1,2 | <span class="built_in">sort</span> | <span class="built_in">uniq</span> &gt; ids.txt</span><br><span class="line"></span><br><span class="line"><span class="comment"># The index name.</span></span><br><span class="line">REF=refs/*.fa</span><br><span class="line"></span><br><span class="line"><span class="comment"># Create the BAM folder.</span></span><br><span class="line"><span class="built_in">mkdir</span> -p bam</span><br><span class="line"></span><br><span class="line"><span class="comment"># Align the FASTQ files to the reference genome.</span></span><br><span class="line"><span class="built_in">cat</span> ids.txt | parallel <span class="string">&quot;hisat2 -x <span class="variable">$&#123;REF&#125;</span> -1 reads/&#123;&#125;_R1.fq -2 reads/&#123;&#125;_R2.fq 2&gt;&gt; alignment_summary.txt | \</span></span><br><span class="line"><span class="string">samtools sort &gt; bam/&#123;&#125;.bam&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Index each BAM file.</span></span><br><span class="line"><span class="built_in">cat</span> ids.txt | parallel  <span class="string">&quot;samtools index bam/&#123;&#125;.bam&quot;</span></span><br></pre></td></tr></table></figure></div><div class="tab-item-content" id="比对-2"><p><code>双端</code>比对信息解读：<br><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">10000 reads; of these:</span><br><span class="line">10000 (100.00%) were paired; of these:</span><br><span class="line">650 (6.50%) aligned concordantly 0 times</span><br><span class="line">8823 (88.23%) aligned concordantly exactly 1 time</span><br><span class="line">527 (5.27%) aligned concordantly &gt;1 times</span><br><span class="line">----</span><br><span class="line">650 pairs aligned concordantly 0 times; of these:</span><br><span class="line">34 (5.23%) aligned discordantly 1 time</span><br><span class="line">----</span><br><span class="line">616 pairs aligned 0 times concordantly or discordantly; of these:</span><br><span class="line">1232 mates make up the pairs; of these:</span><br><span class="line">660 (53.57%) aligned 0 times</span><br><span class="line">571 (46.35%) aligned exactly 1 time</span><br><span class="line">1 (0.08%) aligned &gt;1 times</span><br></pre></td></tr></table></figure></p>
<ul>
<li><p><strong>10000 reads; of these: 10000 (100.00%) were paired</strong>: 总共有10000个读段，所有这些读段都是成对的，意味着每个读段都有相应的配对读段。</p>
</li>
<li><p><strong>650 (6.50%) aligned concordantly 0 times</strong>: 有650对读段（占6.50%）没有一起成功比对到参考序列上。”一致比对”（concordantly）意味着两个读段以预期的相对方向和距离成功地比对到参考序列的同一个区域。</p>
</li>
<li><p><strong>8823 (88.23%) aligned concordantly exactly 1 time</strong>: 有8823对读段（占88.23%）一致地比对到参考序列的唯一位置一次。</p>
</li>
<li><p><strong>527 (5.27%) aligned concordantly &gt;1 times</strong>: 有527对读段（占5.27%）一致地比对到参考序列的多个位置。</p>
</li>
</ul>
<p>接下来的部分提供了关于没有一致比对的读段的更多细节：</p>
<ul>
<li><strong>650 pairs aligned concordantly 0 times; of these:</strong></li>
<li><p><strong>34 (5.23%) aligned discordantly 1 time</strong>: 在没有一致比对的650对读段中，有34对（占5.23%）以不一致的方式比对了一次。”不一致比对”（discordantly）意味着这些读段没有按照预期的相对方向或距离比对到参考序列的同一个区域，但仍然在某处找到了比对位置。</p>
</li>
<li><p><strong>616 pairs aligned 0 times concordantly or discordantly; of these:</strong></p>
</li>
<li><strong>1232 mates make up the pairs; of these:</strong></li>
<li><strong>660 (53.57%) aligned 0 times</strong>: 在这616对既没有一致比对也没有不一致比对的读段中，总共有1232个单独的读段，其中660个（占53.57%）没有比对到参考序列。</li>
<li><strong>571 (46.35%) aligned exactly 1 time</strong>: 有571个读段（占46.35%）比对到了参考序列的唯一位置一次。</li>
<li><strong>1 (0.08%) aligned &gt;1 times</strong>: 有1个读段（占0.08%）比对到了参考序列的多个位置。</li>
</ul>
<p>最后一行提供了总体的比对率：</p>
<ul>
<li><strong>96.70% overall alignment rate</strong>: 总体来说，96.70%的读段至少比对到了参考序列的一个位置。这个百分比是基于所有读段的，包括一致比对的、不一致比对的以及那些只有一个读段比对到参考序列的配对读段。</li>
</ul>
<p><code>单端</code>比对信息解读：<br><figure class="highlight text"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">62364 reads; of these:</span><br><span class="line">62364 (100.00%) were unpaired; of these:</span><br><span class="line">19 (0.03%) aligned 0 times</span><br><span class="line">62311 (99.92%) aligned exactly 1 time</span><br><span class="line">34 (0.05%) aligned &gt;1 times</span><br></pre></td></tr></table></figure></p>
<ul>
<li><p><strong>62364 reads</strong>: 表示总共有62,364个读段(reads)被处理。读段是指从测序实验得到的短序列片段。</p>
</li>
<li><p><strong>of these: 62364 (100.00%) were unpaired</strong>: 所有的读段都是单端读段(unpaired)，意味着这些读段不是成对的。在配对端测序中，你通常会得到两个读段（即一对），它们是从同一个DNA片段的两端测序得到的。单端读段意味着每个片段只测序了一端。</p>
</li>
<li><p><strong>19 (0.03%) aligned 0 times</strong>: 在这些读段中，有19个（占总数的0.03%）没有比对到参考序列上。</p>
</li>
<li><p><strong>62311 (99.92%) aligned exactly 1 time</strong>: 62,311个读段（占总数的99.92%）精确地比对到参考序列上一次。这意味着这些读段在参考序列中有一个唯一的匹配位置。</p>
</li>
<li><p><strong>34 (0.05%) aligned &gt;1 times</strong>: 有34个读段（占总数的0.05%）比对到参考序列上多于一次。这表明这些读段可能来自基因组或转录组中重复区域，因此无法确定它们的精确位置。</p>
</li>
</ul></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div>
<h3 id="bam文件转换"><a href="#bam文件转换" class="headerlink" title="bam文件转换"></a>bam文件转换</h3><p>通常情况下，我们可以将bam文件直接在IGV等可视化软件中查看，但是有时候加载时间很长。当我们只是关注“覆盖率”时，我们可以将文件转换为<code>`bigwig</code>格式，这样加载速度会更快。在这个过程中我们需要将bam文件转换为bedgraph文件，然后再转换为bigwig文件。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># The index name.</span></span><br><span class="line">REF = refs/genome.fa</span><br><span class="line"></span><br><span class="line"><span class="comment"># Turn each BAM file into bedGraph coverage. The files will have the .bg extension.</span></span><br><span class="line"><span class="built_in">cat</span> ids.txt | parallel <span class="string">&quot;bedtools genomecov -ibam  bam/&#123;&#125;.bam -split -bg  &gt; bam/&#123;&#125;.bg&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Convert each bedGraph coverage into bigWig coverage. The files will have the .bw extension.</span></span><br><span class="line"><span class="built_in">cat</span> ids.txt | parallel <span class="string">&quot;bedGraphToBigWig bam/&#123;&#125;.bg <span class="variable">$&#123;REF&#125;</span>.fai bam/&#123;&#125;.bw&quot;</span></span><br></pre></td></tr></table></figure>
<h2 id="计数"><a href="#计数" class="headerlink" title="计数"></a>计数</h2><div class="tabs" id="计数"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="计数-1">featureCounts</button><button type="button" class="tab " data-href="计数-2">featureCounts参数</button><button type="button" class="tab " data-href="计数-3">标准化计数矩阵</button></ul><div class="tab-contents"><div class="tab-item-content active" id="计数-1"><p>得到reads比对文件后，我们需要对比对文件进行计数，得到每个基因的表达量。这里我们可以使用featureCounts工具，featureCounts是Subread软件包的一部分，它可以用于计算基因或转录本的表达量。</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">GTF = refs/*.gtf</span><br><span class="line"></span><br><span class="line"><span class="built_in">cat</span> ids.txt | parallel -j 1 <span class="built_in">echo</span> <span class="string">&quot;bam/&#123;&#125;.bam&quot;</span> | \</span><br><span class="line">    xargs featureCounts -p  --countReadPairs -a <span class="variable">$&#123;GTF&#125;</span> -o counts.txt</span><br></pre></td></tr></table></figure>
<p>此时，<code>parallel</code>，命令参数<code>-j 1</code>是必不可少的，我们要确保命令按照文件中列出的顺序执行。每当顺序至关重要时，您必须使用 -j 1，否则，顺序就无法保证。当是双端测序的时候，我们可以设置<code>-p --countReadPairs</code>参数，这样可以计算每对reads的计数（非必选）。</p></div><div class="tab-item-content" id="计数-2"><p>当我们使用featureCounts时，我们会做出许多决定。除了前面提到的命令外，我们还可以使用其他参数来调整计数的方式。下面是一些常用的参数：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">-t &lt;string&gt;         Specify feature <span class="built_in">type</span>(s) <span class="keyword">in</span> a GTF annotation. If multiple</span><br><span class="line">                      types are provided, they should be separated by <span class="string">&#x27;,&#x27;</span> with</span><br><span class="line">                      no space <span class="keyword">in</span> between. <span class="string">&#x27;exon&#x27;</span> by default. Rows <span class="keyword">in</span> the</span><br><span class="line">                      annotation with a matched feature will be extracted and</span><br><span class="line">                      used <span class="keyword">for</span> <span class="built_in">read</span> mapping. </span><br><span class="line"></span><br><span class="line">-g &lt;string&gt;         Specify attribute <span class="built_in">type</span> <span class="keyword">in</span> GTF annotation. <span class="string">&#x27;gene_id&#x27;</span> by </span><br><span class="line">                      default. Meta-features used <span class="keyword">for</span> <span class="built_in">read</span> counting will be </span><br><span class="line">                      extracted from annotation using the provided value.</span><br><span class="line"></span><br><span class="line">-T &lt;int&gt;            Number of the threads. 1 by default.</span><br></pre></td></tr></table></figure></div><div class="tab-item-content" id="计数-3"><p>生成的counts.txt文件包含了每个基因的表达量信息。可使用自定义脚本<code>format_featurecounts.r</code>来转化为更简单直接的counts数据。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Transform feature counts output to simple counts.</span></span><br><span class="line"><span class="comment"># Adds gene name to a featurecounts output file.</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The design file must have a column named &quot;sample&quot;</span></span><br><span class="line">COUNTS_FILE <span class="operator">=</span> <span class="string">&quot;counts.txt&quot;</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># The output file name.</span></span><br><span class="line">OUTPUT_FILE<span class="operator">=</span> <span class="string">&quot;counts.csv&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the file that contains transcript to gene mapping(for example: human). </span></span><br><span class="line"><span class="comment">## Lists all datasets available in Ensembl</span></span><br><span class="line"><span class="comment">## mart = useMart(&quot;ensembl&quot;, dataset = species_dataset)</span></span><br><span class="line"><span class="comment">## listDatasets(mart)</span></span><br><span class="line"><span class="comment">## Lists all attributes available in datasets</span></span><br><span class="line"><span class="comment">## listAttributes(mart)</span></span><br><span class="line"></span><br><span class="line">SPECIES_DATASET <span class="operator">=</span> <span class="string">&quot;hsapiens_gene_ensembl&quot;</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># Load the required libraries.</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>require<span class="punctuation">(</span><span class="string">&quot;biomaRt&quot;</span><span class="punctuation">,</span> quietly <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;biomaRt&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line">library<span class="punctuation">(</span>biomaRt<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Install the optparse package automatically.</span></span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span>require<span class="punctuation">(</span>optparse<span class="punctuation">,</span> quietly<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  install.packages<span class="punctuation">(</span><span class="string">&quot;optparse&quot;</span><span class="punctuation">,</span> repos<span class="operator">=</span><span class="string">&quot;http://cran.us.r-project.org&quot;</span><span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span>optparse<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Command line options.</span></span><br><span class="line">option_list <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-c&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--counts&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>COUNTS_FILE<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;counts_file&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;input count file  [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-o&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--output_file&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>OUTPUT_FILE<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;output_file&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the design file for the samples [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-s&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--species_dataset&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>SPECIES_DATASET<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;species_dataset&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the transcript to gene mapping [%default]&quot;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Parse the options.</span></span><br><span class="line">opt <span class="operator">=</span> parse_args<span class="punctuation">(</span>OptionParser<span class="punctuation">(</span>option_list<span class="operator">=</span>option_list<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the file that contains the counts.</span></span><br><span class="line">input_file <span class="operator">=</span> opt<span class="operator">$</span>counts_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># The final result file.</span></span><br><span class="line">output_file <span class="operator">=</span> opt<span class="operator">$</span>output_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># The species dataset to gene mapping file.</span></span><br><span class="line">species_dataset <span class="operator">=</span> opt<span class="operator">$</span>species_dataset</span><br><span class="line"></span><br><span class="line"><span class="comment"># Inform the user.</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Reformating featurecounts.\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Input:&quot;</span><span class="punctuation">,</span> input_file<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Read the input data</span></span><br><span class="line">data <span class="operator">=</span> read.table<span class="punctuation">(</span>input_file<span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> check.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> header<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">,</span> sep<span class="operator">=</span><span class="string">&quot;\t&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Remove certain columns.</span></span><br><span class="line">data <span class="operator">=</span> subset<span class="punctuation">(</span>data<span class="punctuation">,</span> select <span class="operator">=</span> <span class="operator">-</span><span class="built_in">c</span><span class="punctuation">(</span><span class="number">2</span><span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">,</span> <span class="number">5</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Remove file related information from the column</span></span><br><span class="line">headers <span class="operator">=</span> sub<span class="punctuation">(</span><span class="string">&quot;\\.bam&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> colnames<span class="punctuation">(</span>data<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">headers <span class="operator">=</span> sub<span class="punctuation">(</span><span class="string">&quot;^(.*)/&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> headers<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Attach the new names</span></span><br><span class="line">colnames<span class="punctuation">(</span>data<span class="punctuation">)</span> <span class="operator">&lt;-</span> headers</span><br><span class="line"></span><br><span class="line"><span class="comment"># Add columns with specific names</span></span><br><span class="line">data<span class="operator">$</span>name <span class="operator">=</span> data<span class="operator">$</span>Geneid</span><br><span class="line">data<span class="operator">$</span>gene <span class="operator">=</span> data<span class="operator">$</span>Geneid</span><br><span class="line"></span><br><span class="line">targets <span class="operator">=</span>  sub<span class="punctuation">(</span><span class="string">&quot;\\..*&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> data<span class="operator">$</span>name<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span>SPECIES_DATASET <span class="operator">!=</span> <span class="string">&#x27;&#x27;</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  cat<span class="punctuation">(</span><span class="string">&quot;# Species_dataset:&quot;</span><span class="punctuation">,</span> SPECIES_DATASET<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Read the transcript to gene mapping file.</span></span><br><span class="line">  mart <span class="operator">=</span> useMart<span class="punctuation">(</span><span class="string">&quot;ensembl&quot;</span><span class="punctuation">,</span> dataset <span class="operator">=</span> SPECIES_DATASET<span class="punctuation">)</span></span><br><span class="line">  tx2gene <span class="operator">=</span> getBM<span class="punctuation">(</span><span class="built_in">attributes</span> <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;ensembl_gene_id&quot;</span><span class="punctuation">,</span> <span class="string">&quot;external_gene_name&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> mart <span class="operator">=</span> mart<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Find the indices where the targets match to the mapping.</span></span><br><span class="line">  indices <span class="operator">=</span> match<span class="punctuation">(</span>targets<span class="punctuation">,</span> tx2gene<span class="operator">$</span>ensembl_gene_id<span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line">  <span class="comment"># Add the gene name column.</span></span><br><span class="line">  data<span class="operator">$</span>gene <span class="operator">=</span> tx2gene<span class="operator">$</span>external_gene_name<span class="punctuation">[</span>indices<span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  data<span class="operator">$</span>gene<span class="punctuation">[</span><span class="built_in">is.na</span><span class="punctuation">(</span>data<span class="operator">$</span>gene<span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">=</span> targets<span class="punctuation">[</span><span class="built_in">is.na</span><span class="punctuation">(</span>data<span class="operator">$</span>gene<span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line">  data<span class="operator">$</span>gene<span class="punctuation">[</span>data<span class="operator">$</span>gene <span class="operator">==</span><span class="string">&quot;&quot;</span><span class="punctuation">]</span> <span class="operator">=</span> targets<span class="punctuation">[</span>data<span class="operator">$</span>gene <span class="operator">==</span><span class="string">&quot;&quot;</span><span class="punctuation">]</span></span><br><span class="line">  </span><br><span class="line"><span class="punctuation">&#125;</span> <span class="keyword">else</span> <span class="punctuation">&#123;</span></span><br><span class="line">  data<span class="operator">$</span>gene <span class="operator">=</span> targets</span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># List the desired column order.</span></span><br><span class="line">cols <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;name&quot;</span><span class="punctuation">,</span> <span class="string">&quot;gene&quot;</span><span class="punctuation">,</span> headers<span class="punctuation">[</span><span class="number">3</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>headers<span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Slice the data.</span></span><br><span class="line">out <span class="operator">=</span> data<span class="punctuation">[</span><span class="punctuation">,</span>cols<span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Save the  resulting summarized counts.</span></span><br><span class="line">write.csv<span class="punctuation">(</span>out<span class="punctuation">,</span> file<span class="operator">=</span>output_file<span class="punctuation">,</span>  row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> <span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Output:&quot;</span><span class="punctuation">,</span> output_file<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div>
<h2 id="差异表达分析"><a href="#差异表达分析" class="headerlink" title="差异表达分析"></a>差异表达分析</h2><div class="tabs" id="差异表达分析"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="差异表达分析-1">差异表达分析</button><button type="button" class="tab " data-href="差异表达分析-2">DESeq2</button></ul><div class="tab-contents"><div class="tab-item-content active" id="差异表达分析-1"><p>得到计数矩阵后我们就可以进行差异表达分析了，常用的差异表达分析工具有DESeq2、edgeR、limma等。</p>
<p>Limma基于线性模型，通过使用贝叶斯方法估计每个基因的差异方差。它使用经验贝叶斯方法来将信息从所有基因中借用，特别是在样本较少时提高估计的稳定性。<br>edgeR基于负二项分布模型。它使用贝叶斯方法通过适应组内变异估计提高估计的稳定性。edgeR考虑了基因的丰度和变异性，使其更适用于RNA-Seq数据。<br>DESeq2基于负二项分布的模型。它通过使用贝叶斯方法来考虑样本间的差异以及基因表达的离散性。</p></div><div class="tab-item-content" id="差异表达分析-2"><p>这里我们使用DESeq2进行差异表达分析，我们需要用到两个文件，一个是<code>counts.csv</code>，且是我们在上一步骤简化过的；另一个是<code>design.csv</code>，它包含了实验设计信息，一列是<code>sample</code>，一列是<code>group</code>，group是我们要比较的分组信息。</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Differential expression analysis with the DESeq2 package.</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># https://bioconductor.org/packages/release/bioc/html/DESeq2.html</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Command line install: mamba install bioconductor-deseq2</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"></span><br><span class="line">DESIGN_FILE <span class="operator">=</span> <span class="string">&quot;design.csv&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the file that contains the counts.</span></span><br><span class="line">COUNTS_FILE <span class="operator">=</span> <span class="string">&quot;counts.csv&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The output file.</span></span><br><span class="line">OUTPUT_FILE <span class="operator">=</span> <span class="string">&quot;deseq2.csv&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Install the optparse package automatically.</span></span><br><span class="line"><span class="keyword">if</span><span class="punctuation">(</span><span class="operator">!</span>require<span class="punctuation">(</span>optparse<span class="punctuation">,</span> quietly<span class="operator">=</span><span class="built_in">T</span><span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  install.packages<span class="punctuation">(</span><span class="string">&quot;optparse&quot;</span><span class="punctuation">,</span> repos<span class="operator">=</span><span class="string">&quot;http://cran.us.r-project.org&quot;</span><span class="punctuation">)</span></span><br><span class="line">  library<span class="punctuation">(</span>optparse<span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Command line options.</span></span><br><span class="line">option_list <span class="operator">=</span> <span class="built_in">list</span><span class="punctuation">(</span> </span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-d&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--design_file&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>DESIGN_FILE<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;design_file&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the design file for the samples [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-c&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--counts&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>COUNTS_FILE<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;counts_file&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;input counts file  [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-o&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--out&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span>OUTPUT_FILE<span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;output_file&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the  results file  [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-f&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--factor_name&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span><span class="string">&quot;group&quot;</span><span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;factor_name&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the factor column name in the design file [%default]&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">  </span><br><span class="line">  make_option<span class="punctuation">(</span><span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;-s&quot;</span><span class="punctuation">,</span> <span class="string">&quot;--sample_name&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> action<span class="operator">=</span><span class="string">&quot;store&quot;</span><span class="punctuation">,</span> default<span class="operator">=</span><span class="string">&quot;sample&quot;</span><span class="punctuation">,</span> dest<span class="operator">=</span><span class="string">&quot;sample_name&quot;</span><span class="punctuation">,</span> type<span class="operator">=</span><span class="string">&#x27;character&#x27;</span><span class="punctuation">,</span></span><br><span class="line">              help<span class="operator">=</span><span class="string">&quot;the sample column name in the design file [%default]&quot;</span><span class="punctuation">)</span></span><br><span class="line">  </span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Parse the options.</span></span><br><span class="line">opt <span class="operator">=</span> parse_args<span class="punctuation">(</span>OptionParser<span class="punctuation">(</span>option_list<span class="operator">=</span>option_list<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the file that contains the counts.</span></span><br><span class="line">counts_file <span class="operator">=</span> opt<span class="operator">$</span>counts_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># The sample file is in CSV format and must have the headers &quot;sample&quot; and &quot;condition&quot;.</span></span><br><span class="line">design_file <span class="operator">=</span> opt<span class="operator">$</span>design_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># The output file.</span></span><br><span class="line">output_file <span class="operator">=</span> opt<span class="operator">$</span>output_file</span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the factor column</span></span><br><span class="line">factor_name <span class="operator">=</span> opt<span class="operator">$</span>factor_name</span><br><span class="line"></span><br><span class="line"><span class="comment"># The name of the sample column</span></span><br><span class="line">sample_name <span class="operator">=</span> opt<span class="operator">$</span>sample_name</span><br><span class="line"></span><br><span class="line"><span class="comment"># Inform the user.</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Running DESeq2&quot;</span><span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Design:&quot;</span><span class="punctuation">,</span> design_file<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Counts:&quot;</span><span class="punctuation">,</span> counts_file<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Read the design file.</span></span><br><span class="line">inp <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span>design_file<span class="punctuation">,</span> strip.white <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Check the sample column name.</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Sample column:&quot;</span><span class="punctuation">,</span> sample_name<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>sample_name <span class="operator">%in%</span> colnames<span class="punctuation">(</span>inp<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  stop<span class="punctuation">(</span><span class="string">&quot;# Sample column: &quot;</span><span class="punctuation">,</span> sample_name<span class="punctuation">,</span> <span class="string">&quot; not found in design file (use -s).&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Check the factor column name.</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Factor column:&quot;</span><span class="punctuation">,</span> factor_name<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="keyword">if</span> <span class="punctuation">(</span><span class="operator">!</span>factor_name <span class="operator">%in%</span> colnames<span class="punctuation">(</span>inp<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  stop<span class="punctuation">(</span><span class="string">&quot;# Factor column: &quot;</span><span class="punctuation">,</span> factor_name<span class="punctuation">,</span> <span class="string">&quot; not found in design file (use -f).&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a new data frame to work with.</span></span><br><span class="line">df <span class="operator">=</span> data.frame<span class="punctuation">(</span></span><br><span class="line">  sample <span class="operator">=</span> inp<span class="punctuation">[</span><span class="punctuation">,</span> sample_name<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">  groups <span class="operator">=</span> inp<span class="punctuation">[</span><span class="punctuation">,</span> factor_name<span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The default groups come from the design</span></span><br><span class="line">groups <span class="operator">=</span> unique<span class="punctuation">(</span>df<span class="operator">$</span>group<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a new data frame to work with.</span></span><br><span class="line">df <span class="operator">=</span> data.frame<span class="punctuation">(</span></span><br><span class="line">  sample <span class="operator">=</span> inp<span class="punctuation">[</span><span class="punctuation">,</span> sample_name<span class="punctuation">]</span><span class="punctuation">,</span></span><br><span class="line">  groups <span class="operator">=</span> inp<span class="punctuation">[</span><span class="punctuation">,</span> factor_name<span class="punctuation">]</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The default groups come from the design</span></span><br><span class="line">groups <span class="operator">=</span> unique<span class="punctuation">(</span>df<span class="operator">$</span>groups <span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Print information on the groups</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>grp <span class="keyword">in</span> groups<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">  n_grp <span class="operator">&lt;-</span> <span class="built_in">sum</span><span class="punctuation">(</span>df<span class="operator">$</span>groups <span class="operator">==</span> grp<span class="punctuation">)</span></span><br><span class="line">  cat<span class="punctuation">(</span><span class="string">&quot;# Group&quot;</span><span class="punctuation">,</span> grp<span class="punctuation">,</span> <span class="string">&quot;has&quot;</span><span class="punctuation">,</span> n_grp<span class="punctuation">,</span> <span class="string">&quot;samples.\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Turn conditions into factors.</span></span><br><span class="line">df<span class="operator">$</span>group <span class="operator">=</span> factor<span class="punctuation">(</span>df<span class="operator">$</span>group<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The first level should correspond to the first entry in the file</span></span><br><span class="line"><span class="comment"># Required later when building a model.</span></span><br><span class="line">df<span class="operator">$</span>group <span class="operator">=</span> relevel<span class="punctuation">(</span>df<span class="operator">$</span>group<span class="punctuation">,</span> toString<span class="punctuation">(</span>df<span class="operator">$</span>group<span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Isolate the sample names.</span></span><br><span class="line">sample_names <span class="operator">=</span> df<span class="operator">$</span>sample</span><br><span class="line"></span><br><span class="line"><span class="comment"># Read the data from input file.</span></span><br><span class="line">count_df <span class="operator">=</span> read.csv<span class="punctuation">(</span>counts_file<span class="punctuation">,</span> strip.white <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">,</span> stringsAsFactors <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Created rounded integers for the count data</span></span><br><span class="line">count_data <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>count_df<span class="punctuation">[</span><span class="punctuation">,</span> sample_names<span class="punctuation">,</span> drop<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Column names should be carried over. Skip sample names and FDR</span></span><br><span class="line">other_cols <span class="operator">=</span> <span class="built_in">names</span><span class="punctuation">(</span>count_df<span class="punctuation">)</span><span class="punctuation">[</span><span class="operator">!</span><span class="built_in">names</span><span class="punctuation">(</span>count_df<span class="punctuation">)</span> <span class="operator">%in%</span> <span class="built_in">c</span><span class="punctuation">(</span>sample_names<span class="punctuation">,</span> <span class="string">&quot;FDR&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Carry over all additional columns not in sample names.</span></span><br><span class="line">row_data <span class="operator">=</span> count_df<span class="punctuation">[</span><span class="punctuation">,</span> other_cols<span class="punctuation">,</span> drop<span class="operator">=</span><span class="built_in">F</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># Running DESeq2</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Bioconductor packages used in the script.</span></span><br><span class="line">bio.p <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;DESeq2&quot;</span><span class="punctuation">,</span> <span class="string">&quot;tibble&quot;</span><span class="punctuation">,</span> <span class="string">&quot;dplyr&quot;</span><span class="punctuation">,</span> <span class="string">&quot;tools&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Load Bioconductor packages</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Initializing &quot;</span><span class="punctuation">,</span> bio.p<span class="punctuation">,</span> <span class="string">&quot;...&quot;</span><span class="punctuation">)</span></span><br><span class="line">status <span class="operator">=</span> suppressPackageStartupMessages<span class="punctuation">(</span></span><br><span class="line">  lapply<span class="punctuation">(</span>bio.p<span class="punctuation">,</span> library<span class="punctuation">,</span> character.only <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot; done\n&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create DESEq2 dataset.</span></span><br><span class="line">dds <span class="operator">=</span> DESeqDataSetFromMatrix<span class="punctuation">(</span></span><br><span class="line">  countData <span class="operator">=</span> count_data<span class="punctuation">,</span></span><br><span class="line">  rowData <span class="operator">=</span> row_data<span class="punctuation">,</span></span><br><span class="line">  colData <span class="operator">=</span> df<span class="punctuation">,</span></span><br><span class="line">  design <span class="operator">=</span> <span class="operator">~</span>group</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Number of rows before</span></span><br><span class="line">n_start <span class="operator">=</span> nrow<span class="punctuation">(</span>dds<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Remove rows with no counts</span></span><br><span class="line">keep <span class="operator">&lt;-</span> rowSums<span class="punctuation">(</span>counts<span class="punctuation">(</span>dds<span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">&gt;</span> <span class="number">3</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># At least 3 samples with a count of 10 or higher</span></span><br><span class="line"><span class="comment"># keep &lt;- rowSums(counts(dds) &gt;= 10) &gt;= 3</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Apply the filtering</span></span><br><span class="line">dds <span class="operator">&lt;-</span> dds<span class="punctuation">[</span>keep<span class="punctuation">,</span> <span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Run deseq</span></span><br><span class="line">dds <span class="operator">=</span> DESeq<span class="punctuation">(</span>dds<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Get the normalized counts.</span></span><br><span class="line">normed <span class="operator">=</span> counts<span class="punctuation">(</span>dds<span class="punctuation">,</span> normalized<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Round normalized counts to a single digit.</span></span><br><span class="line">normed <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>normed<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Format the results.</span></span><br><span class="line">res <span class="operator">=</span> results<span class="punctuation">(</span>dds<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"><span class="comment"># The rest of the code is about formatting the output dataframe.</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The first columns that carry over from the input.</span></span><br><span class="line">start <span class="operator">=</span>  dplyr<span class="operator">::</span>as_tibble<span class="punctuation">(</span>rowData<span class="punctuation">(</span>dds<span class="punctuation">)</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span> other_cols<span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># The main tibble</span></span><br><span class="line">data <span class="operator">=</span> bind_cols<span class="punctuation">(</span>start<span class="punctuation">,</span> as_tibble<span class="punctuation">(</span>res<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create the foldChange column.</span></span><br><span class="line">data<span class="operator">$</span>foldChange <span class="operator">=</span> <span class="number">2</span> <span class="operator">^</span> data<span class="operator">$</span>log2FoldChange</span><br><span class="line"></span><br><span class="line"><span class="comment"># Rename the columns.</span></span><br><span class="line">data <span class="operator">=</span>  dplyr<span class="operator">::</span>rename<span class="punctuation">(</span>data<span class="punctuation">,</span> PValue<span class="operator">=</span>pvalue<span class="punctuation">,</span> FDR<span class="operator">=</span>padj<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Set all NA values to 1.</span></span><br><span class="line">data<span class="operator">$</span>FDR<span class="punctuation">[</span><span class="built_in">is.na</span><span class="punctuation">(</span>data<span class="operator">$</span>FDR<span class="punctuation">)</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> <span class="number">1</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a real adjusted pvalue</span></span><br><span class="line">data<span class="operator">$</span>PAdj <span class="operator">=</span> p.adjust<span class="punctuation">(</span>data<span class="operator">$</span>PValue<span class="punctuation">,</span> method<span class="operator">=</span><span class="string">&quot;hochberg&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create the additional columns that we wish to present.</span></span><br><span class="line">data<span class="operator">$</span>baseMeanA <span class="operator">=</span> <span class="number">1</span></span><br><span class="line">data<span class="operator">$</span>baseMeanB <span class="operator">=</span> <span class="number">1</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Combine the normed matrix with the results</span></span><br><span class="line">total <span class="operator">&lt;-</span> bind_cols<span class="punctuation">(</span>data<span class="punctuation">,</span> normed<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Sort again by P-value.</span></span><br><span class="line">total <span class="operator">=</span> arrange<span class="punctuation">(</span>total<span class="punctuation">,</span> FDR<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Compute the false discovery counts on the sorted table.</span></span><br><span class="line">total<span class="operator">$</span>falsePos <span class="operator">=</span> <span class="number">1</span><span class="operator">:</span>nrow<span class="punctuation">(</span>total<span class="punctuation">)</span> <span class="operator">*</span> total<span class="operator">$</span>FDR</span><br><span class="line"></span><br><span class="line"><span class="comment"># Sample names for condition A and B</span></span><br><span class="line">col_names_A <span class="operator">&lt;-</span> sample_names <span class="punctuation">[</span>df<span class="operator">$</span>group <span class="operator">==</span> groups<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line">col_names_B <span class="operator">&lt;-</span> sample_names <span class="punctuation">[</span>df<span class="operator">$</span>group <span class="operator">==</span> groups<span class="punctuation">[[</span><span class="number">2</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#col_names_A = unlist(df %&gt;% filter(group==groups[1]) %&gt;% select(sample))</span></span><br><span class="line"><span class="comment">#col_names_B = unlist(df %&gt;% filter(group==groups[2]) %&gt;% select(sample))</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create the individual baseMean columns.</span></span><br><span class="line">total<span class="operator">$</span>baseMeanA <span class="operator">=</span> rowMeans<span class="punctuation">(</span>total<span class="punctuation">[</span><span class="punctuation">,</span> col_names_A<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>baseMeanB <span class="operator">=</span> rowMeans<span class="punctuation">(</span>total<span class="punctuation">[</span><span class="punctuation">,</span> col_names_B<span class="punctuation">]</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Bringing some sanity to numbers. Round columns to fewer digits.</span></span><br><span class="line">total<span class="operator">$</span>foldChange <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>foldChange<span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>log2FoldChange <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>log2FoldChange<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>baseMean  <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>baseMean<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>baseMeanA <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>baseMeanA<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>baseMeanB <span class="operator">=</span>  <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>baseMeanB<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>lfcSE <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>lfcSE<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>stat <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>stat<span class="punctuation">,</span> <span class="number">2</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>FDR <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>FDR<span class="punctuation">,</span> <span class="number">4</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>falsePos <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>total<span class="operator">$</span>falsePos<span class="punctuation">,</span> <span class="number">0</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Reorganize columns names to make more sense.</span></span><br><span class="line">new_cols <span class="operator">=</span> <span class="built_in">c</span><span class="punctuation">(</span>other_cols<span class="punctuation">,</span> </span><br><span class="line">             <span class="string">&quot;baseMean&quot;</span><span class="punctuation">,</span><span class="string">&quot;baseMeanA&quot;</span><span class="punctuation">,</span><span class="string">&quot;baseMeanB&quot;</span><span class="punctuation">,</span></span><br><span class="line">             <span class="string">&quot;foldChange&quot;</span><span class="punctuation">,</span> <span class="string">&quot;log2FoldChange&quot;</span><span class="punctuation">,</span></span><br><span class="line">             <span class="string">&quot;lfcSE&quot;</span><span class="punctuation">,</span><span class="string">&quot;stat&quot;</span><span class="punctuation">,</span><span class="string">&quot;PValue&quot;</span><span class="punctuation">,</span><span class="string">&quot;PAdj&quot;</span><span class="punctuation">,</span> </span><br><span class="line">             <span class="string">&quot;FDR&quot;</span><span class="punctuation">,</span><span class="string">&quot;falsePos&quot;</span><span class="punctuation">,</span> col_names_A<span class="punctuation">,</span> col_names_B<span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Slice the dataframe with new columns.</span></span><br><span class="line">total <span class="operator">=</span> total<span class="punctuation">[</span><span class="punctuation">,</span> new_cols<span class="punctuation">]</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Count the number of significant results.</span></span><br><span class="line">n_total <span class="operator">=</span> nrow<span class="punctuation">(</span>total<span class="punctuation">)</span></span><br><span class="line">fdr_count <span class="operator">=</span> <span class="built_in">sum</span><span class="punctuation">(</span>total<span class="operator">$</span>FDR <span class="operator">&lt;</span> <span class="number">0.05</span><span class="punctuation">,</span> na.rm<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">fdr_frac <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>fdr_count<span class="operator">/</span>n_total<span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span> <span class="operator">*</span> <span class="number">100</span></span><br><span class="line"></span><br><span class="line">pval_count <span class="operator">=</span> <span class="built_in">sum</span><span class="punctuation">(</span>total<span class="operator">$</span>PValue <span class="operator">&lt;</span> <span class="number">0.05</span><span class="punctuation">,</span> na.rm<span class="operator">=</span><span class="literal">TRUE</span><span class="punctuation">)</span></span><br><span class="line">pval_frac <span class="operator">=</span> <span class="built_in">round</span><span class="punctuation">(</span>pval_count<span class="operator">/</span>n_total<span class="punctuation">,</span> <span class="number">3</span><span class="punctuation">)</span> <span class="operator">*</span> <span class="number">100</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Reformat these columns as string.</span></span><br><span class="line">total<span class="operator">$</span>PAdj <span class="operator">=</span> formatC<span class="punctuation">(</span>total<span class="operator">$</span>PAdj<span class="punctuation">,</span> format <span class="operator">=</span> <span class="string">&quot;e&quot;</span><span class="punctuation">,</span> digits <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line">total<span class="operator">$</span>PValue <span class="operator">=</span> formatC<span class="punctuation">(</span>total<span class="operator">$</span>PValue<span class="punctuation">,</span> format <span class="operator">=</span> <span class="string">&quot;e&quot;</span><span class="punctuation">,</span> digits <span class="operator">=</span> <span class="number">1</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Write the results to the standard output.</span></span><br><span class="line">write.csv<span class="punctuation">(</span>total<span class="punctuation">,</span> file<span class="operator">=</span>output_file<span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> <span class="built_in">quote</span> <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Inform the user.</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Input:&quot;</span><span class="punctuation">,</span> n_start<span class="punctuation">,</span> <span class="string">&quot;rows\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Removed:&quot;</span><span class="punctuation">,</span> n_start<span class="operator">-</span>n_total<span class="punctuation">,</span> <span class="string">&quot;rows\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Fitted:&quot;</span><span class="punctuation">,</span> n_total<span class="punctuation">,</span> <span class="string">&quot;rows\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Significant PVal:&quot;</span><span class="punctuation">,</span> pval_count<span class="punctuation">,</span> <span class="string">&quot;(&quot;</span><span class="punctuation">,</span> pval_frac<span class="punctuation">,</span> <span class="string">&quot;%)\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Significant FDRs:&quot;</span><span class="punctuation">,</span> fdr_count<span class="punctuation">,</span> <span class="string">&quot;(&quot;</span><span class="punctuation">,</span> fdr_frac<span class="punctuation">,</span> <span class="string">&quot;%)\n&quot;</span><span class="punctuation">)</span></span><br><span class="line">cat<span class="punctuation">(</span><span class="string">&quot;# Results:&quot;</span><span class="punctuation">,</span> output_file<span class="punctuation">,</span> <span class="string">&quot;\n&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div> 
<h1 id="RNA-Seq分析所遇问题"><a href="#RNA-Seq分析所遇问题" class="headerlink" title="RNA-Seq分析所遇问题"></a>RNA-Seq分析所遇问题</h1><h2 id="比对问题"><a href="#比对问题" class="headerlink" title="比对问题"></a>比对问题</h2><p>编码序列的GC含量（指基因序列中鸟嘌呤（G）和胞嘧啶（C）的比例）通常高于整个基因组的平均GC含量。这是因为编码序列往往需要维持一定的稳定性，GC对之间的氢键比AT对更多，增加了DNA双螺旋的稳定性。在基因组测序和比对分析中，如果一个区域的GC含量远离平均水平，不管是偏高还是偏低，都可能导致测序和分析上的偏差。例如，高GC含量的区域可能因为PCR扩增偏好性或测序平台的偏好性而过度表示。此外，极端的GC含量可能影响测序读段的比对，因为这样的区域在基因组中可能不是唯一的，或者比对算法在处理这些区域时可能不够准确。</p>
<p>查看序列GC含量：<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看序列GC含量</span></span><br><span class="line"><span class="built_in">cat</span> *.fa | seqkit <span class="built_in">seq</span> -s | geecee --filter</span><br><span class="line">samtools faidx *.fa 1 |seqkit <span class="built_in">seq</span> -s | geecee --filter</span><br><span class="line"><span class="built_in">cat</span> *.fq | seqkit <span class="built_in">seq</span> -s | geecee --filter</span><br></pre></td></tr></table></figure></p>
<p>特别的，当我们发现RNA-Seq数据比对出现长内含子区域时，可能就是由于参考序列中较长的仅有<code>T</code>组成造成的，在HISAT2中，我们可以使用<code>--max-intronlen</code>参数来调整最大内含子长度，以提高比对正确率。经典的内含子长度为500bp，我们可以将这个数值设置为2500.<br><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Make a directory for the bam file.</span></span><br><span class="line"><span class="built_in">mkdir</span> -p bam</span><br><span class="line"></span><br><span class="line"><span class="comment"># Run the hisat aligner for each sample.</span></span><br><span class="line"><span class="built_in">cat</span> ids | parallel <span class="string">&quot;hisat2 --max-intronlen 2500 -x <span class="variable">$IDX</span> -U reads/&#123;&#125;.fq  | samtools sort &gt; bam/&#123;&#125;.bam&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Create a bam index for each file.</span></span><br><span class="line"><span class="built_in">cat</span> ids | parallel samtools index bam/&#123;&#125;.bam</span><br></pre></td></tr></table></figure></p>
<h2 id="计数问题"><a href="#计数问题" class="headerlink" title="计数问题"></a>计数问题</h2><p>  在使用featureCount时，默认使用exon进行计数，通常有两种策略，一种是在基因水平分析，一种是在转录本水平分析。在基因水平分析时，外显子只会被计数一次，而在转录本水平时，外显子可能出现在多个转录本中，如果简单的每个外显子计数一次，可能会导致计数的重复，导致从贡献度被过度估计。为了解决这个问题，需要一个过程来确定每个外显子对每个转录本的贡献比例。这个过程称为“解卷积”（deconvolution）。解卷积的目的是分配每个外显子的表达量，确保即使一个外显子在多个转录本中出现，它的总贡献也只被计算一次。解卷积通常涉及到数学和统计方法，可以基于转录本的预期丰度或其他生物信息学数据来估计每个外显子的相对贡献。<br>  一般情况下，建议使用featureCounts进行基因水平分析，使用基于分类/伪比对的方法如kallisto和salmon等工具进行转录水平分析。</p>
<h2 id="链特异性问题"><a href="#链特异性问题" class="headerlink" title="链特异性问题"></a>链特异性问题</h2><p>  RNA-Seq 文库类型的选择决定了测序的读取方向和 cDNA 链的测序顺序，这意味着来自不同文库类型的 RNA-Seq 读取可能存在显著差异。有关文库类型的信息对于将读取组装成转录组或映射到参考组装非常有用。这是因为文库类型可以帮助通过读取的相对方向和从哪条链测序来帮助辨别转录组中较短的模糊读取属于哪里。不幸的是，有关所用文库类型的信息未包含在测序输出文件中，并且可能在数据组装之前丢失。即使使用来自公共存储库的 RNA-Seq 数据，也不能保证文库类型信息是正确的或根本不存在。<br>  在成对端测序中，每对测序读取包括一个正向（forward）和一个反向（reverse）组分。这些读取对应于 DNA 片段的两端，通常用于提高测序的覆盖率和准确性。然而，当数据是成对的且有方向性时，理解每个读取是如何对应于参考基因组或转录本的变得更加复杂。在有方向性的成对端对齐中，通常第一对中的读取（first in pair）会与反义链（antisense strand）对齐，而第二对中的读取（second in pair）则会与正义链（sense strand）对齐。例如，如果 DNA 来自于反向链（reverse strand）上的一个转录本，那么第一对中的读取会对齐到正向链（forward strand）上，并且会在参考序列上向左移动（leftward shift）；第二对中的读取会对齐到反向链，并且相对于片段会向右移动（rightward shift）。虽然这听起来“简单”，实际上却不是这样。在测序数据的方向性问题上存在很多混淆和误解，以至于有专门的软件工具，例如“<a target="_blank" rel="noopener" href="https://github.com/NBISweden/GUESSmyLT">Guess My Library Type</a>”，来帮助研究人员弄清楚他们的数据是否有方向性，以及是哪种类型的方向性。</p>
<p>  在一些软件中，有参数可以用来指定来链的<a target="_blank" rel="noopener" href="https://github.com/igordot/genomics/blob/master/notes/rna-seq-strand.md">方向</a>。</p>
<h2 id="转录完整性"><a href="#转录完整性" class="headerlink" title="转录完整性"></a>转录完整性</h2><h1 id="RNA-Seq分析流程"><a href="#RNA-Seq分析流程" class="headerlink" title="RNA-Seq分析流程"></a>RNA-Seq分析流程</h1><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649ce4ac2c1fe.png" alt="分析流程.png"></p>
<div class="tabs" id="分析流程"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="分析流程-1">比对和组装</button><button type="button" class="tab " data-href="分析流程-2">定量和标准化</button><button type="button" class="tab " data-href="分析流程-3">差异分析</button></ul><div class="tab-contents"><div class="tab-item-content active" id="分析流程-1"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649ce8628dc1e.png" alt="比对和组装.png"></p></div><div class="tab-item-content" id="分析流程-2"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649ce991c1407.png" alt="定量和标准化.png"></p></div><div class="tab-item-content" id="分析流程-3"><p><strong><a target="_blank" rel="noopener" href="https://www.smileszh.cn/posts/1085fed1.html">差异分析</a></strong></p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649cea75e840e.png" alt="差异分析.png"></p>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649d900223506.png" alt="基因过滤.png"></p></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div>
<h1 id="基因表达量数据类型"><a href="#基因表达量数据类型" class="headerlink" title="基因表达量数据类型"></a>基因表达量数据类型</h1><div class="tabs" id="数据类型"><ul class="nav-tabs"><button type="button" class="tab  active" data-href="数据类型-1">Counts</button><button type="button" class="tab " data-href="数据类型-2">RPM/CPM</button><button type="button" class="tab " data-href="数据类型-3">RPKM</button><button type="button" class="tab " data-href="数据类型-4">FPKM/FPKM-UQ</button><button type="button" class="tab " data-href="数据类型-5">TPM</button></ul><div class="tab-contents"><div class="tab-item-content active" id="数据类型-1"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e3cab4acae.png" alt="Counts.png"></p></div><div class="tab-item-content" id="数据类型-2"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e3d78cb226.png" alt="RPM.png"></p></div><div class="tab-item-content" id="数据类型-3"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e3db0533fb.png" alt="RPKM.png"></p></div><div class="tab-item-content" id="数据类型-4"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e3ea1474b5.png" alt="FPKM:FPKM-UQ.png"></p></div><div class="tab-item-content" id="数据类型-5"><p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e3efc298af.png" alt="TPM.png"></p></div></div><div class="tab-to-top"><button type="button" aria-label="scroll to top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div>
<h1 id="基因表达量数据转换"><a href="#基因表达量数据转换" class="headerlink" title="基因表达量数据转换"></a>基因表达量数据转换</h1><ul>
<li>原始Counts数据</li>
<li>基因长度信息（外显子长度之和）</li>
<li>如按照protein-coding基因计算mapped reads计算，还需提供基因的分类信息</li>
</ul>
<p><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/30/649e40f51a9cd.png" alt="基因表达量数据转换.png"></p>
</article><div class="post-copyright"><div class="copyright-cc-box"><i class="anzhiyufont anzhiyu-icon-copyright"></i></div><div class="post-copyright__author_box"><a class="post-copyright__author_img" href="/" title="头像"><img class="post-copyright__author_img_back" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/07/05/64a521d49890b.gif" title="头像" alt="头像"><img class="post-copyright__author_img_front" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/07/05/64a521d49890b.gif" title="头像" alt="头像"></a><div class="post-copyright__author_name">泡泡</div><div class="post-copyright__author_desc">平安喜乐，万事胜意。</div></div><div class="post-copyright__post__info"><a class="post-copyright__original" title="该文章为原创文章，注意版权协议" href="https://smileszh.github.io/posts/7acd0447.html">原创</a><a class="post-copyright-title"><span onclick="rm.copyPageUrl('https://smileszh.github.io/posts/7acd0447.html')">RNA-Seq</span></a></div><div class="post-tools" id="post-tools"><div class="post-tools-left"><div class="rewardLeftButton"></div><div class="shareRight"><div class="share-link mobile"><div class="share-qrcode"><div class="share-button" title="使用手机访问这篇文章"><i class="anzhiyufont anzhiyu-icon-qrcode"></i></div><div class="share-main"><div class="share-main-all"><div id="qrcode" title="https://smileszh.github.io/posts/7acd0447.html"></div><div class="reward-dec">使用手机访问这篇文章</div></div></div></div></div><div class="share-link weibo"><a class="share-button" target="_blank" href="https://service.weibo.com/share/share.php?title=RNA-Seq&amp;url=https://smileszh.github.io/posts/7acd0447.html&amp;pic=https://bu.dusays.com/2023/07/03/64a283d519f00.png" rel="external nofollow noreferrer noopener"><i class="anzhiyufont anzhiyu-icon-weibo"></i></a></div><div class="share-link copyurl"><div class="share-button" id="post-share-url" title="复制链接" onclick="rm.copyPageUrl()"><i class="anzhiyufont anzhiyu-icon-link"></i></div></div></div></div></div><div class="post-copyright__notice"><span class="post-copyright-info">本博客所有文章除特别声明外，均采用 <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/" target="_blank">CC BY-NC-SA 4.0</a> 许可协议。转载请注明来自 <a href="https://smileszh.github.io" target="_blank">海底捞里没有鱼</a>！</span></div></div><div class="post-tools-right"><div class="tag_share"><div class="post-meta__box"><div class="post-meta__box__tag-list"><a class="post-meta__box__tags" href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/"><span class="tags-punctuation"> <i class="anzhiyufont anzhiyu-icon-tag"></i></span>转录组<span class="tagsPageCount">1</span></a></div></div></div><div class="post_share"><div class="social-share" data-image="https://bu.dusays.com/2024/05/29/6656fab1234ae.png" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.cbd.int/butterfly-extsrc@1.1.3/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"/><script src="https://cdn.cbd.int/butterfly-extsrc@1.1.3/sharejs/dist/js/social-share.min.js" defer="defer"></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/posts/53f493a2.html"><img class="prev-cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/28/649c429eb975a.jpg" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="pagination-info"><div class="label">上一篇</div><div class="prev_info">疑惑当铺</div></div></a></div><div class="next-post pull-right"><a href="/posts/795e1915.html"><img class="next-cover" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/29/649d446fa97cc.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="pagination-info"><div class="label">下一篇</div><div class="next_info">生信数据格式</div></div></a></div></nav><hr/><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="anzhiyufont anzhiyu-icon-comments"></i><span> 评论</span></div><div class="comment-randomInfo"><a onclick="anzhiyu.addRandomCommentInfo()" href="javascript:void(0)">匿名评论</a><a href="/privacy" style="margin-left: 4px">隐私政策</a></div><div class="comment-tips" id="comment-tips"><span>✅ 你无需删除空行，直接评论以获取最佳展示效果</span></div></div><div class="comment-wrap"><div><div id="twikoo-wrap"></div></div></div></div><div class="comment-barrage"></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="card-content"><div class="author-info__sayhi" id="author-info__sayhi" onclick="anzhiyu.changeSayHelloText()"></div><div class="author-info-avatar"><img class="avatar-img" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/07/05/64a521d49890b.gif" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/><div class="author-status"><img class="g-status" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/08/24/64e6ce9c507bb.png" ait="status"/></div></div><div class="author-info__description">⛰️愿我们永远生活在自由中，被自由之风推动着。</div><div class="author-info__bottom-group"><a class="author-info__bottom-group-left" href="/"><h1 class="author-info__name">泡泡</h1><div class="author-info__desc">平安喜乐，万事胜意。</div></a><div class="card-info-social-icons is-center"><a class="social-icon faa-parent animated-hover" href="https://github.com/smileszh" target="_blank" title="Github"><i class="anzhiyufont anzhiyu-icon-github"></i></a><a class="social-icon faa-parent animated-hover" href="mailto:1976267283@qq.com" target="_blank" title="Email"><i class="anzhiyufont anzhiyu-icon-envelope"></i></a><a class="social-icon faa-parent animated-hover" href="/atom.xml" target="_blank" title="RSS"><i class="anzhiyufont anzhiyu-icon-rss"></i></a><a class="social-icon faa-parent animated-hover" href="https://space.bilibili.com/358266510" target="_blank" title="BiliBili"><i class="anzhiyufont anzhiyu-icon-bilibili"></i></a><a class="social-icon faa-parent animated-hover" href="https://bu.dusays.com/2023/06/06/647f506fa8b99.png" target="_blank" title="QQ"><i class="anzhiyufont anzhiyu-icon-qq"></i></a></div></div></div></div><div class="card-widget anzhiyu-right-widget" id="card-wechat" onclick="null"><div id="flip-wrapper"><div id="flip-content"><div class="face" style="background: url(https://bu.dusays.com/2023/06/06/647f504132f84.png) center center / 100% no-repeat"></div><div class="back face" style="background: url(https://bu.dusays.com/2023/06/06/647f503ebaae4.png) center center / 100% no-repeat"></div></div></div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-bars"></i><span>文章目录</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-1"><a class="toc-link" href="#RNA-Seq%E6%8C%87%E5%8D%97"><span class="toc-number">1.</span> <span class="toc-text">RNA-Seq指南</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#RNA-Seq%E5%8F%82%E8%80%83"><span class="toc-number">1.1.</span> <span class="toc-text">RNA-Seq参考</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E5%9F%BA%E5%9B%A0%E7%BB%84%E8%BF%9B%E8%A1%8C%E9%87%8F%E5%8C%96"><span class="toc-number">1.1.1.</span> <span class="toc-text">根据基因组进行量化</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%A0%B9%E6%8D%AE%E8%BD%AC%E5%BD%95%E7%BB%84%E8%BF%9B%E8%A1%8C%E5%88%86%E7%B1%BB"><span class="toc-number">1.1.2.</span> <span class="toc-text">根据转录组进行分类</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84"><span class="toc-number">1.2.</span> <span class="toc-text">参考基因组</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%A4%A7%E5%B0%8F"><span class="toc-number">1.2.1.</span> <span class="toc-text">参考基因组大小</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9C%89%E5%A4%9A%E5%B0%91%E4%B8%AA%E6%9F%93%E8%89%B2%E4%BD%93"><span class="toc-number">1.2.2.</span> <span class="toc-text">有多少个染色体</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%AF%8F%E4%B8%AA%E6%9F%93%E8%89%B2%E4%BD%93%E7%9A%84%E9%95%BF%E5%BA%A6"><span class="toc-number">1.2.3.</span> <span class="toc-text">每个染色体的长度</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6"><span class="toc-number">1.3.</span> <span class="toc-text">参考基因组注释文件</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%BA%86%E8%A7%A3%E6%95%B0%E6%8D%AE"><span class="toc-number">1.4.</span> <span class="toc-text">了解数据</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9"><span class="toc-number">1.5.</span> <span class="toc-text">比对</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%9E%84%E5%BB%BA%E7%B4%A2%E5%BC%95"><span class="toc-number">1.5.1.</span> <span class="toc-text">构建索引</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9-1"><span class="toc-number">1.5.2.</span> <span class="toc-text">比对</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#bam%E6%96%87%E4%BB%B6%E8%BD%AC%E6%8D%A2"><span class="toc-number">1.5.3.</span> <span class="toc-text">bam文件转换</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%AE%A1%E6%95%B0"><span class="toc-number">1.6.</span> <span class="toc-text">计数</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90"><span class="toc-number">1.7.</span> <span class="toc-text">差异表达分析</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#RNA-Seq%E5%88%86%E6%9E%90%E6%89%80%E9%81%87%E9%97%AE%E9%A2%98"><span class="toc-number">2.</span> <span class="toc-text">RNA-Seq分析所遇问题</span></a><ol class="toc-child"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%AF%94%E5%AF%B9%E9%97%AE%E9%A2%98"><span class="toc-number">2.1.</span> <span class="toc-text">比对问题</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%AE%A1%E6%95%B0%E9%97%AE%E9%A2%98"><span class="toc-number">2.2.</span> <span class="toc-text">计数问题</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E9%93%BE%E7%89%B9%E5%BC%82%E6%80%A7%E9%97%AE%E9%A2%98"><span class="toc-number">2.3.</span> <span class="toc-text">链特异性问题</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%BD%AC%E5%BD%95%E5%AE%8C%E6%95%B4%E6%80%A7"><span class="toc-number">2.4.</span> <span class="toc-text">转录完整性</span></a></li></ol></li><li class="toc-item toc-level-1"><a class="toc-link" href="#RNA-Seq%E5%88%86%E6%9E%90%E6%B5%81%E7%A8%8B"><span class="toc-number">3.</span> <span class="toc-text">RNA-Seq分析流程</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E9%87%8F%E6%95%B0%E6%8D%AE%E7%B1%BB%E5%9E%8B"><span class="toc-number">4.</span> <span class="toc-text">基因表达量数据类型</span></a></li><li class="toc-item toc-level-1"><a class="toc-link" href="#%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E9%87%8F%E6%95%B0%E6%8D%AE%E8%BD%AC%E6%8D%A2"><span class="toc-number">5.</span> <span class="toc-text">基因表达量数据转换</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="anzhiyufont anzhiyu-icon-history"></i><span>最近发布</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/posts/d08e2cc.html" title="Biostar workflows"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/06/11/6667bb0ee5f60.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Biostar workflows"/></a><div class="content"><a class="title" href="/posts/d08e2cc.html" title="Biostar workflows">Biostar workflows</a><time datetime="2024-06-11T03:00:00.000Z" title="发表于 2024-06-11 11:00:00">2024-06-11</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/posts/b536050b.html" title="Makefile基本使用"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/05/29/6656fab1234ae.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Makefile基本使用"/></a><div class="content"><a class="title" href="/posts/b536050b.html" title="Makefile基本使用">Makefile基本使用</a><time datetime="2024-05-29T10:00:00.000Z" title="发表于 2024-05-29 18:00:00">2024-05-29</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/posts/ca675cd4.html" title="Git使用"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/05/02/663358f92bc74.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Git使用"/></a><div class="content"><a class="title" href="/posts/ca675cd4.html" title="Git使用">Git使用</a><time datetime="2024-05-02T10:00:00.000Z" title="发表于 2024-05-02 18:00:00">2024-05-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/posts/16088ba7.html" title="bookdown使用"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2024/04/30/6630e3274908b.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="bookdown使用"/></a><div class="content"><a class="title" href="/posts/16088ba7.html" title="bookdown使用">bookdown使用</a><time datetime="2024-04-30T14:00:00.000Z" title="发表于 2024-04-30 22:00:00">2024-04-30</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/posts/2ea551a.html" title="ggplot2包"><img src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/08/27/64ea25faa9244.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="ggplot2包"/></a><div class="content"><a class="title" href="/posts/2ea551a.html" title="ggplot2包">ggplot2包</a><time datetime="2023-08-08T00:00:00.000Z" title="发表于 2023-08-08 08:00:00">2023-08-08</time></div></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div id="footer_deal"><a class="deal_link" href="mailto:1976267283@qq.com" title="email"><i class="anzhiyufont anzhiyufont anzhiyu-icon-envelope"></i></a><a class="deal_link" target="_blank" rel="noopener" href="https://weibo.com/u/6638725378" title="微博"><i class="anzhiyufont anzhiyufont anzhiyu-icon-weibo"></i></a><a class="deal_link" href="/null" title="facebook"><i class="anzhiyufont anzhiyufont anzhiyu-icon-facebook1"></i></a><img class="footer_mini_logo" title="返回顶部" alt="返回顶部" onclick="anzhiyu.scrollToDest(0, 500)" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/07/05/64a521d49890b.gif" size="50px"/><a class="deal_link" target="_blank" rel="noopener" href="https://github.com/smileszh" title="Github"><i class="anzhiyufont anzhiyufont anzhiyu-icon-github"></i></a><a class="deal_link" target="_blank" rel="noopener" href="https://space.bilibili.com/358266510?spm_id_from=333.1007.0.0" title="Bilibili"><i class="anzhiyufont anzhiyufont anzhiyu-icon-bilibili"></i></a><a class="deal_link" href="/null" title="抖音"><i class="anzhiyufont anzhiyufont anzhiyu-icon-tiktok"></i></a></div><div class="copyright">&copy;2023 - 2024 By 泡泡</div><div id="workboard"><img class="workSituationImg boardsign" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://img.shields.io/badge/%E5%B0%8F%E8%8B%8F-%E4%B8%8A%E7%8F%AD%E6%91%B8%E9%B1%BC%E4%B8%AD-9cf?style=social&amp;logo=gnometerminal" alt="上班努力摸鱼中~" title="上班努力摸鱼中~"/><div id="runtimeTextTip"></div></div></div></footer></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="sidebar-site-data site-data is-center"><a href="/archives/" title="archive"><div class="headline">文章</div><div class="length-num">23</div></a><a href="/tags/" title="tag"><div class="headline">标签</div><div class="length-num">23</div></a><a href="/categories/" title="category"><div class="headline">分类</div><div class="length-num">4</div></a></div><span class="sidebar-menu-item-title">功能</span><div class="sidebar-menu-item"><a class="darkmode_switchbutton menu-child" href="javascript:void(0);" title="显示模式"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i><span>显示模式</span></a></div><div class="back-menu-list-groups"><div class="back-menu-list-group"><div class="back-menu-list-title">网页</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://smileszh.cn" title="博客首页"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="/img/favicon.ico" alt="博客首页"/><span class="back-menu-item-text">博客首页</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">项目</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://github.com/smileszh/smileszh.github.io" title="Github"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/07/648036b97aaba.png" alt="Github"/><span class="back-menu-item-text">Github</span></a><a class="back-menu-item" target="_blank" rel="noopener" href="http://127.0.0.1:5244/" title="alist"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/07/12/64adad75d8996.png" alt="alist"/><span class="back-menu-item-text">alist</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">图床</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://7bu.top/" title="去不图床"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/07/6480380fd3f0b.png" alt="去不图床"/><span class="back-menu-item-text">去不图床</span></a><a class="back-menu-item" target="_blank" rel="noopener" href="https://image.anheyu.com/" title="安和鱼图床"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/07/6480380fd3f0b.png" alt="安和鱼图床"/><span class="back-menu-item-text">安和鱼图床</span></a></div></div><div class="back-menu-list-group"><div class="back-menu-list-title">云盘</div><div class="back-menu-list"><a class="back-menu-item" target="_blank" rel="noopener" href="https://yun.console.xlj0.com/login?redirect=%2Flogin" title="Tooc公益网盘"><img class="back-menu-item-icon" src= "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7" onerror="this.onerror=null,this.src=&quot;/img/404.jpg&quot;" data-lazy-src="https://bu.dusays.com/2023/06/20/6490a64ff04d3.png" alt="Tooc公益网盘"/><span class="back-menu-item-text">Tooc公益网盘</span></a></div></div></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 文章</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/archives/"><i class="anzhiyufont anzhiyu-icon-box-archive faa-tada" style="font-size: 0.9em;"></i><span> 隧道</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/categories/"><i class="anzhiyufont anzhiyu-icon-shapes faa-tada" style="font-size: 0.9em;"></i><span> 分类</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags faa-tada" style="font-size: 0.9em;"></i><span> 标签</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 友链</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/link/"><i class="anzhiyufont anzhiyu-icon-link faa-tada" style="font-size: 0.9em;"></i><span> 友人帐</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/fcircle/"><i class="anzhiyufont anzhiyu-icon-artstation faa-tada" style="font-size: 0.9em;"></i><span> 朋友圈</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/comments/"><i class="anzhiyufont anzhiyu-icon-envelope faa-tada" style="font-size: 0.9em;"></i><span> 留言板</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 生活</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/music/"><i class="anzhiyufont anzhiyu-icon-music faa-tada" style="font-size: 0.9em;"></i><span> 音乐馆</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/bangumis/"><i class="anzhiyufont anzhiyu-icon-bilibili faa-tada" style="font-size: 0.9em;"></i><span> 追番页</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/album/"><i class="anzhiyufont anzhiyu-icon-images faa-tada" style="font-size: 0.9em;"></i><span> 相册集</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="javascript:void(0);"><span> 我的</span></a><ul class="menus_item_child"><li><a class="site-page child faa-parent animated-hover" href="/about/"><i class="anzhiyufont anzhiyu-icon-paper-plane faa-tada" style="font-size: 0.9em;"></i><span> 关于我们</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/essay/"><i class="anzhiyufont anzhiyu-icon-lightbulb faa-tada" style="font-size: 0.9em;"></i><span> 闲言碎语</span></a></li><li><a class="site-page child faa-parent animated-hover" href="/equipment/"><i class="anzhiyufont anzhiyu-icon-dice-d20 faa-tada" style="font-size: 0.9em;"></i><span> 我的装备</span></a></li></ul></div></div><span class="sidebar-menu-item-title">标签</span><div class="card-tags"><div class="item-headline"></div><div class="card-tag-cloud"><a href="/tags/GO-KEGG/" style="font-size: 0.88rem;">GO/KEGG<sup>2</sup></a><a href="/tags/Mac/" style="font-size: 0.88rem;">Mac<sup>1</sup></a><a href="/tags/R-%E8%AF%AD%E8%A8%80/" style="font-size: 0.88rem;">R 语言<sup>1</sup></a><a href="/tags/R%E5%8C%85/" style="font-size: 0.88rem;">R包<sup>4</sup></a><a href="/tags/R%E7%BB%98%E5%9B%BE/" style="font-size: 0.88rem;">R绘图<sup>2</sup></a><a href="/tags/R%E8%AF%AD%E8%A8%80/" style="font-size: 0.88rem;">R语言<sup>9</sup></a><a href="/tags/Terminal/" style="font-size: 0.88rem;">Terminal<sup>1</sup></a><a href="/tags/conda/" style="font-size: 0.88rem;">conda<sup>1</sup></a><a href="/tags/tidyverse/" style="font-size: 0.88rem;">tidyverse<sup>1</sup></a><a href="/tags/%E5%88%86%E5%AD%90%E4%BA%92%E4%BD%9C/" style="font-size: 0.88rem;">分子互作<sup>1</sup></a><a href="/tags/%E5%9F%BA%E5%9B%A0ID/" style="font-size: 0.88rem;">基因ID<sup>2</sup></a><a href="/tags/%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A/" style="font-size: 0.88rem;">基因注释<sup>2</sup></a><a href="/tags/%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/" style="font-size: 0.88rem;">富集分析<sup>1</sup></a><a href="/tags/%E5%B7%AE%E5%BC%82%E8%A1%A8%E6%B3%95%E5%88%86%E6%9E%90/" style="font-size: 0.88rem;">差异表法分析<sup>1</sup></a><a href="/tags/%E5%BC%95%E7%89%A9%E8%AE%BE%E8%AE%A1/" style="font-size: 0.88rem;">引物设计<sup>1</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" style="font-size: 0.88rem;">数据分析<sup>1</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 0.88rem;">数据库<sup>3</sup></a><a href="/tags/%E6%95%B0%E6%8D%AE%E6%A0%BC%E5%BC%8F/" style="font-size: 0.88rem;">数据格式<sup>1</sup></a><a href="/tags/%E7%94%9F%E7%89%A9%E6%95%B0%E6%8D%AE%E5%BA%93/" style="font-size: 0.88rem;">生物数据库<sup>1</sup></a><a href="/tags/%E7%9F%A5%E8%AF%86%E7%82%B9/" style="font-size: 0.88rem;">知识点<sup>1</sup></a><a href="/tags/%E8%81%9A%E7%B1%BB%E5%88%86%E6%9E%90/" style="font-size: 0.88rem;">聚类分析<sup>1</sup></a><a href="/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" style="font-size: 0.88rem;">转录组<sup>1</sup></a><a href="/tags/%E8%BD%AF%E4%BB%B6%E7%AE%A1%E7%90%86/" style="font-size: 0.88rem;">软件管理<sup>1</sup></a></div></div><hr/></div></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="阅读模式"><i class="anzhiyufont anzhiyu-icon-book-open"></i></button><button id="translateLink" type="button" title="简繁转换">繁</button><button id="darkmode" type="button" title="浅色和深色模式转换"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i></button><button id="hide-aside-btn" type="button" title="单栏和双栏切换"><i class="anzhiyufont anzhiyu-icon-arrows-left-right"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="设置"><i class="anzhiyufont anzhiyu-icon-gear"></i></button><button class="close" id="mobile-toc-button" type="button" title="目录"><i class="anzhiyufont anzhiyu-icon-list-ul"></i></button><a id="to_comment" href="#post-comment" title="直达评论"><i class="anzhiyufont anzhiyu-icon-comments"></i></a><a id="switch-commentBarrage" href="javascript:anzhiyu.switchCommentBarrage();" title="开关弹幕"><i class="anzhiyufont anzhiyu-icon-danmu"></i></a><button id="go-up" type="button" title="回到顶部"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></button></div></div><div id="nav-music"><a id="nav-music-hoverTips" onclick="anzhiyu.musicToggle()" accesskey="m">播放音乐</a><div id="console-music-bg"></div><meting-js id="8152976493" server="netease" type="playlist" mutex="true" preload="none" theme="var(--anzhiyu-main)" data-lrctype="0" order="random"></meting-js></div><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">搜索</span><span id="loading-status"></span><button class="search-close-button"><i class="anzhiyufont anzhiyu-icon-xmark"></i></button></nav><div class="is-center" id="loading-database"><i class="anzhiyufont anzhiyu-icon-spinner anzhiyu-pulse-icon"></i><span>  数据库加载中</span></div><div class="search-wrap"><div id="local-search-input"><div class="local-search-box"><input class="local-search-box--input" placeholder="搜索文章" type="text"/></div></div><hr/><div id="local-search-results"></div></div></div><div id="search-mask"></div></div><div id="rightMenu"><div class="rightMenu-group rightMenu-small"><div class="rightMenu-item" id="menu-backward"><i class="anzhiyufont anzhiyu-icon-arrow-left"></i></div><div class="rightMenu-item" id="menu-forward"><i class="anzhiyufont anzhiyu-icon-arrow-right"></i></div><div class="rightMenu-item" id="menu-refresh"><i class="anzhiyufont anzhiyu-icon-arrow-rotate-right" style="font-size: 1rem;"></i></div><div class="rightMenu-item" id="menu-top"><i class="anzhiyufont anzhiyu-icon-arrow-up"></i></div></div><div class="rightMenu-group rightMenu-line rightMenuPlugin"><div class="rightMenu-item" id="menu-copytext"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制选中文本</span></div><div class="rightMenu-item" id="menu-pastetext"><i class="anzhiyufont anzhiyu-icon-paste"></i><span>粘贴文本</span></div><a class="rightMenu-item" id="menu-commenttext"><i class="anzhiyufont anzhiyu-icon-comment-medical"></i><span>引用到评论</span></a><div class="rightMenu-item" id="menu-newwindow"><i class="anzhiyufont anzhiyu-icon-window-restore"></i><span>新窗口打开</span></div><div class="rightMenu-item" id="menu-copylink"><i class="anzhiyufont anzhiyu-icon-link"></i><span>复制链接地址</span></div><div class="rightMenu-item" id="menu-copyimg"><i class="anzhiyufont anzhiyu-icon-images"></i><span>复制此图片</span></div><div class="rightMenu-item" id="menu-downloadimg"><i class="anzhiyufont anzhiyu-icon-download"></i><span>下载此图片</span></div><div class="rightMenu-item" id="menu-newwindowimg"><i class="anzhiyufont anzhiyu-icon-window-restore"></i><span>新窗口打开图片</span></div><div class="rightMenu-item" id="menu-search"><i class="anzhiyufont anzhiyu-icon-magnifying-glass"></i><span>站内搜索</span></div><div class="rightMenu-item" id="menu-searchBaidu"><i class="anzhiyufont anzhiyu-icon-magnifying-glass"></i><span>百度搜索</span></div><div class="rightMenu-item" id="menu-music-toggle"><i class="anzhiyufont anzhiyu-icon-play"></i><span>播放音乐</span></div><div class="rightMenu-item" id="menu-music-back"><i class="anzhiyufont anzhiyu-icon-backward"></i><span>切换到上一首</span></div><div class="rightMenu-item" id="menu-music-forward"><i class="anzhiyufont anzhiyu-icon-forward"></i><span>切换到下一首</span></div><div class="rightMenu-item" id="menu-music-playlist" onclick="window.open(&quot;https://y.qq.com/n/ryqq/playlist/8802438608&quot;, &quot;_blank&quot;);" style="display: none;"><i class="anzhiyufont anzhiyu-icon-radio"></i><span>查看所有歌曲</span></div><div class="rightMenu-item" id="menu-music-copyMusicName"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制歌名</span></div></div><div class="rightMenu-group rightMenu-line rightMenuOther"><a class="rightMenu-item menu-link" id="menu-randomPost"><i class="anzhiyufont anzhiyu-icon-shuffle"></i><span>随便逛逛</span></a><a class="rightMenu-item menu-link" href="/categories/"><i class="anzhiyufont anzhiyu-icon-cube"></i><span>博客分类</span></a><a class="rightMenu-item menu-link" href="/tags/"><i class="anzhiyufont anzhiyu-icon-tags"></i><span>文章标签</span></a></div><div class="rightMenu-group rightMenu-line rightMenuOther"><a class="rightMenu-item" id="menu-copy" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-copy"></i><span>复制地址</span></a><a class="rightMenu-item" id="menu-commentBarrage" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-message"></i><span class="menu-commentBarrage-text">关闭热评</span></a><a class="rightMenu-item" id="menu-darkmode" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-circle-half-stroke"></i><span class="menu-darkmode-text">深色模式</span></a><a class="rightMenu-item" id="menu-translate" href="javascript:void(0);"><i class="anzhiyufont anzhiyu-icon-language"></i><span>轉為繁體</span></a></div></div><div id="rightmenu-mask"></div><div id="he-plugin-simple"></div><script>var WIDGET = {
  "CONFIG": {
    "modules": "0124",
    "background": "2",
    "tmpColor": "FFFFFF",
    "tmpSize": "16",
    "cityColor": "FFFFFF",
    "citySize": "16",
    "aqiColor": "E8D87B",
    "aqiSize": "16",
    "weatherIconSize": "24",
    "alertIconSize": "18",
    "padding": "10px 10px 10px 10px",
    "shadow": "0",
    "language": "auto",
    "borderRadius": "20",
    "fixed": "true",
    "vertical": "top",
    "horizontal": "left",
    "left": "20",
    "top": "7.1",
    "key": "df245676fb434a0691ead1c63341cd94"
  }
}
</script><link rel="stylesheet" href="https://widget.qweather.net/simple/static/css/he-simple.css?v=1.4.0"/><script src="https://widget.qweather.net/simple/static/js/he-simple.js?v=1.4.0"></script><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="/js/tw_cn.js"></script><script src="https://cdn.cbd.int/@fancyapps/ui@5.0.20/dist/fancybox/fancybox.umd.js"></script><script src="https://cdn.cbd.int/instant.page@5.2.0/instantpage.js" type="module"></script><script src="https://cdn.cbd.int/vanilla-lazyload@17.8.4/dist/lazyload.iife.min.js"></script><script src="https://cdn.cbd.int/node-snackbar@0.1.16/dist/snackbar.min.js"></script><canvas id="universe"></canvas><script async src="https://npm.elemecdn.com/anzhiyu-theme-static@1.0.0/dark/dark.js"></script><script>// 消除控制台打印
var HoldLog = console.log;
console.log = function () {};
let now1 = new Date();
queueMicrotask(() => {
  const Log = function () {
    HoldLog.apply(console, arguments);
  }; //在恢复前输出日志
  const grt = new Date("06/01/2023 00:00:00"); //此处修改你的建站时间或者网站上线时间
  now1.setTime(now1.getTime() + 250);
  const days = (now1 - grt) / 1000 / 60 / 60 / 24;
  const dnum = Math.floor(days);
  const ascll = [
    `欢迎使用安知鱼!`,
    `生活明朗, 万物可爱`,
    `
        
       █████╗ ███╗   ██╗███████╗██╗  ██╗██╗██╗   ██╗██╗   ██╗
      ██╔══██╗████╗  ██║╚══███╔╝██║  ██║██║╚██╗ ██╔╝██║   ██║
      ███████║██╔██╗ ██║  ███╔╝ ███████║██║ ╚████╔╝ ██║   ██║
      ██╔══██║██║╚██╗██║ ███╔╝  ██╔══██║██║  ╚██╔╝  ██║   ██║
      ██║  ██║██║ ╚████║███████╗██║  ██║██║   ██║   ╚██████╔╝
      ╚═╝  ╚═╝╚═╝  ╚═══╝╚══════╝╚═╝  ╚═╝╚═╝   ╚═╝    ╚═════╝
        
        `,
    "已上线",
    dnum,
    "天",
    "©2023 By 安知鱼 V1.6.4",
  ];
  const ascll2 = [`NCC2-036`, `调用前置摄像头拍照成功，识别为【小笨蛋】.`, `Photo captured: `, `🤪`];

  setTimeout(
    Log.bind(
      console,
      `\n%c${ascll[0]} %c ${ascll[1]} %c ${ascll[2]} %c${ascll[3]}%c ${ascll[4]}%c ${ascll[5]}\n\n%c ${ascll[6]}\n`,
      "color:#3b70fc",
      "",
      "color:#3b70fc",
      "color:#3b70fc",
      "",
      "color:#3b70fc",
      ""
    )
  );
  setTimeout(
    Log.bind(
      console,
      `%c ${ascll2[0]} %c ${ascll2[1]} %c \n${ascll2[2]} %c\n${ascll2[3]}\n`,
      "color:white; background-color:#4fd953",
      "",
      "",
      'background:url("https://npm.elemecdn.com/anzhiyu-blog@1.1.6/img/post/common/tinggge.gif") no-repeat;font-size:450%'
    )
  );

  setTimeout(Log.bind(console, "%c WELCOME %c 你好，小笨蛋.", "color:white; background-color:#4f90d9", ""));

  setTimeout(
    console.warn.bind(
      console,
      "%c ⚡ Powered by 安知鱼 %c 你正在访问 泡泡 的博客.",
      "color:white; background-color:#f0ad4e",
      ""
    )
  );

  setTimeout(Log.bind(console, "%c W23-12 %c 你已打开控制台.", "color:white; background-color:#4f90d9", ""));

  setTimeout(
    console.warn.bind(console, "%c S013-782 %c 你现在正处于监控中.", "color:white; background-color:#d9534f", "")
  );
});</script><script async src="/anzhiyu/random.js"></script><script async="async">(function () {
  var grt = new Date("06/01/2023 00:00:00"); //设置网站上线时间
  var now = new Date();
  var dnum;
  var hnum;
  var mnum;
  var snum;
  var nowHour;

  // 计算并更新天数、小时数、分钟数和秒数
  function updateTime() {
    now = new Date(); // 更新 now 的值
    nowHour = now.getHours(); // 更新 nowHour 的值
    var days = (now - grt) / 1000 / 60 / 60 / 24;
    dnum = Math.floor(days);
    var hours = (now - grt) / 1000 / 60 / 60 - 24 * dnum;
    hnum = Math.floor(hours);
    if (String(hnum).length == 1) {
      hnum = "0" + hnum;
    }
    var minutes = (now - grt) / 1000 / 60 - 24 * 60 * dnum - 60 * hnum;
    mnum = Math.floor(minutes);
    if (String(mnum).length == 1) {
      mnum = "0" + mnum;
    }
    var seconds = (now - grt) / 1000 - 24 * 60 * 60 * dnum - 60 * 60 * hnum - 60 * mnum;
    snum = Math.round(seconds);
    if (String(snum).length == 1) {
      snum = "0" + snum;
    }
  }

  // 更新网页中显示的网站运行时间
  function updateHtml() {
    const footer = document.getElementById("footer");
    if (!footer) return
    let currentTimeHtml = "";
    if (nowHour < 18 && nowHour >= 9) {
      // 如果是上班时间，默认就是"安知鱼-上班摸鱼中.svg"图片，不需要更改
      currentTimeHtml = `本站居然运行了 ${dnum} 天<span id='runtime'> ${hnum} 小时 ${mnum} 分 ${snum} 秒 </span><i class='anzhiyufont anzhiyu-icon-heartbeat' style='color:red'></i>`;
    } else {
      // 如果是下班时间，插入"安知鱼-下班啦.svg"图片
      let img = document.querySelector("#workboard .workSituationImg");
      img.src = "https://img.shields.io/badge/%E5%B0%8F%E8%8B%8F-%E4%B8%8B%E7%8F%AD%E5%95%A6%EF%BD%9E-brightgreen?style=social&amp;logo=aseprite&amp;label=%E5%B0%8F%E8%8B%8F";
      img.title = "下班了就该开开心心的玩耍，嘿嘿~";
      img.alt = "下班了就该开开心心的玩耍，嘿嘿~";

      currentTimeHtml = `本站居然运行了 ${dnum} 天<span id='runtime'> ${hnum} 小时 ${mnum} 分 ${snum} 秒 </span><i class='anzhiyufont anzhiyu-icon-heartbeat' style='color:red'></i>`;
    }

    if (document.getElementById("runtimeTextTip")) {
      document.getElementById("runtimeTextTip").innerHTML = currentTimeHtml;
    }
  }

  setInterval(() => {
    updateTime();
    updateHtml();
  }, 1000);
})();</script><script src="/js/search/local-search.js"></script><div class="js-pjax"><script>(() => {
  const init = () => {
    twikoo.init(Object.assign({
      el: '#twikoo-wrap',
      envId: 'https://twikoo.smileszh.cn/',
      region: '',
      onCommentLoaded: () => {
        anzhiyu.loadLightbox(document.querySelectorAll('#twikoo .tk-content img:not(.tk-owo-emotion)'))
      }
    }, null))
  }

  const loadTwikoo = () => {
    if (typeof twikoo === 'object') setTimeout(runFn,0)
    else getScript('https://cdn.cbd.int/twikoo@1.6.17/dist/twikoo.all.min.js').then(runFn)
  }

  const getCount = () => {
    const countELement = document.getElementById('twikoo-count')
    if(!countELement) return
    twikoo.getCommentsCount({
      envId: 'https://twikoo.smileszh.cn/',
      region: '',
      urls: [window.location.pathname],
      includeReply: false
    }).then(res => {
      countELement.textContent = res[0].count
    }).catch(err => {
      console.error(err)
    })
  }

  const runFn = () => {
    init();
    
  }

  if ('Twikoo' === 'Twikoo' || !true) {
    if (true) anzhiyu.loadComment(document.getElementById('twikoo-wrap'), loadTwikoo)
    else {
      loadTwikoo()
    }
  } else {
    window.loadOtherComment = loadTwikoo
  }
})()</script><input type="hidden" name="page-type" id="page-type" value="post"><script async src="/js/anzhiyu/comment_barrage.js"></script></div><script>window.addEventListener('load', () => {
  const changeContent = (content) => {
    if (content === '') return content

    content = content.replace(/<img.*?src="(.*?)"?[^\>]+>/ig, '[图片]') // replace image link
    content = content.replace(/<a[^>]+?href=["']?([^"']+)["']?[^>]*>([^<]+)<\/a>/gi, '[链接]') // replace url
    content = content.replace(/<pre><code>.*?<\/pre>/gi, '[代码]') // replace code
    content = content.replace(/<[^>]+>/g,"") // remove html tag

    if (content.length > 150) {
      content = content.substring(0,150) + '...'
    }
    return content
  }

  const getComment = () => {
    const runTwikoo = () => {
      twikoo.getRecentComments({
        envId: 'https://twikoo.smileszh.cn/',
        region: '',
        pageSize: 6,
        includeReply: true
      }).then(function (res) {
        const twikooArray = res.map(e => {
          return {
            'content': changeContent(e.comment),
            'avatar': e.avatar,
            'nick': e.nick,
            'url': e.url + '#' + e.id,
            'date': new Date(e.created).toISOString()
          }
        })

        saveToLocal.set('twikoo-newest-comments', JSON.stringify(twikooArray), 10/(60*24))
        generateHtml(twikooArray)
      }).catch(function (err) {
        const $dom = document.querySelector('#card-newest-comments .aside-list')
        $dom.textContent= "无法获取评论，请确认相关配置是否正确"
      })
    }

    if (typeof twikoo === 'object') {
      runTwikoo()
    } else {
      getScript('https://cdn.cbd.int/twikoo@1.6.17/dist/twikoo.all.min.js').then(runTwikoo)
    }
  }

  const generateHtml = array => {
    let result = ''

    if (array.length) {
      for (let i = 0; i < array.length; i++) {
        result += '<div class=\'aside-list-item\'>'

        if (true) {
          const name = 'data-lazy-src'
          result += `<a href='${array[i].url}' class='thumbnail'><img ${name}='${array[i].avatar}' alt='${array[i].nick}'><div class='name'><span>${array[i].nick} </span></div></a>`
        }
        
        result += `<div class='content'>
        <a class='comment' href='${array[i].url}' title='${array[i].content}'>${array[i].content}</a>
        <time datetime="${array[i].date}">${anzhiyu.diffDate(array[i].date, true)}</time></div>
        </div>`
      }
    } else {
      result += '没有评论'
    }

    let $dom = document.querySelector('#card-newest-comments .aside-list')
    $dom.innerHTML= result
    window.lazyLoadInstance && window.lazyLoadInstance.update()
    window.pjax && window.pjax.refresh($dom)
  }

  const newestCommentInit = () => {
    if (document.querySelector('#card-newest-comments .aside-list')) {
      const data = saveToLocal.get('twikoo-newest-comments')
      if (data) {
        generateHtml(JSON.parse(data))
      } else {
        getComment()
      }
    }
  }

  newestCommentInit()
  document.addEventListener('pjax:complete', newestCommentInit)
})</script><script async data-pjax src="https://npm.elemecdn.com/anzhiyu-theme-static@1.0.1/bubble/bubble.js"></script><script>var visitorMail = "visitor@anheyu.com";
</script><script async data-pjax src="https://cdn.cbd.int/anzhiyu-theme-static@1.0.0/waterfall/waterfall.js"></script><script src="https://lf3-cdn-tos.bytecdntp.com/cdn/expire-1-M/qrcodejs/1.0.0/qrcode.min.js"></script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.1.9/icon/ali_iconfont_css.css"><script defer="defer" id="fluttering_ribbon" mobile="false" src="https://cdn.cbd.int/butterfly-extsrc@1.1.3/dist/canvas-fluttering-ribbon.min.js"></script><script src="https://cdn.cbd.int/butterfly-extsrc@1.1.3/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = false;
document.body.addEventListener('input', POWERMODE);
</script><link rel="stylesheet" href="https://cdn.cbd.int/anzhiyu-theme-static@1.0.0/aplayer/APlayer.min.css" media="print" onload="this.media='all'"><script src="https://cdn.cbd.int/anzhiyu-blog-static@1.0.1/js/APlayer.min.js"></script><script src="https://cdn.cbd.int/hexo-anzhiyu-music@1.0.1/assets/js/Meting2.min.js"></script><script src="https://cdn.cbd.int/pjax@0.2.8/pjax.min.js"></script><script>let pjaxSelectors = ["meta[property=\"og:image\"]","meta[property=\"og:title\"]","meta[property=\"og:url\"]","meta[property=\"og:type\"]","meta[property=\"og:site_name\"]","meta[property=\"og:description\"]","head > title","#config-diff","#body-wrap","#rightside-config-hide","#rightside-config-show",".js-pjax"]
var pjax = new Pjax({
  elements: 'a:not([target="_blank"])',
  selectors: pjaxSelectors,
  cacheBust: false,
  analytics: false,
  scrollRestoration: false
})

document.addEventListener('pjax:send', function () {
  // removeEventListener scroll 
  anzhiyu.removeGlobalFnEvent('pjax')
  anzhiyu.removeGlobalFnEvent('themeChange')

  document.getElementById('rightside').classList.remove('rightside-show')
  
  if (window.aplayers) {
    for (let i = 0; i < window.aplayers.length; i++) {
      if (!window.aplayers[i].options.fixed) {
        window.aplayers[i].destroy()
      }
    }
  }

  typeof typed === 'object' && typed.destroy()

  //reset readmode
  const $bodyClassList = document.body.classList
  $bodyClassList.contains('read-mode') && $bodyClassList.remove('read-mode')
})

document.addEventListener('pjax:complete', function () {
  window.refreshFn()

  document.querySelectorAll('script[data-pjax]').forEach(item => {
    const newScript = document.createElement('script')
    const content = item.text || item.textContent || item.innerHTML || ""
    Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
    newScript.appendChild(document.createTextNode(content))
    item.parentNode.replaceChild(newScript, item)
  })

  GLOBAL_CONFIG.islazyload && window.lazyLoadInstance.update()

  typeof chatBtnFn === 'function' && chatBtnFn()
  typeof panguInit === 'function' && panguInit()

  // google analytics
  typeof gtag === 'function' && gtag('config', '', {'page_path': window.location.pathname});

  // baidu analytics
  typeof _hmt === 'object' && _hmt.push(['_trackPageview',window.location.pathname]);

  typeof loadMeting === 'function' && document.getElementsByClassName('aplayer').length && loadMeting()

  // prismjs
  typeof Prism === 'object' && Prism.highlightAll()
})

document.addEventListener('pjax:error', e => {
  if (e.request.status === 404) {
    pjax.loadUrl('/404.html')
  }
})</script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><script charset="UTF-8" src="https://cdn.cbd.int/anzhiyu-theme-static@1.1.5/accesskey/accesskey.js"></script></div><div id="popup-window"><div class="popup-window-title">通知</div><div class="popup-window-divider"></div><div class="popup-window-content"><div class="popup-tip">你好呀</div><div class="popup-link"><i class="anzhiyufont anzhiyu-icon-arrow-circle-right"></i></div></div></div><script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"tagMode":false,"debug":false,"model":{"jsonPath":"/live2dw/assets/koharu.model.json"},"display":{"position":"left","width":150,"height":300},"mobile":{"show":false},"log":false,"pluginJsPath":"lib/","pluginModelPath":"assets/","pluginRootPath":"live2dw/"});</script></body></html>